// @mastra/core/dist/provider-registry-EHOAWHFE.cjs
'use strict';

var chunkOWZ6QT24_cjs = require('./chunk-OWZ6QT24.cjs');



Object.defineProperty(exports, "GatewayRegistry", {
  enumerable: true,
  get: function () { return chunkOWZ6QT24_cjs.GatewayRegistry; }
});
Object.defineProperty(exports, "PROVIDER_MODELS", {
  enumerable: true,
  get: function () { return chunkOWZ6QT24_cjs.PROVIDER_MODELS; }
});
Object.defineProperty(exports, "PROVIDER_REGISTRY", {
  enumerable: true,
  get: function () { return chunkOWZ6QT24_cjs.PROVIDER_REGISTRY; }
});
Object.defineProperty(exports, "getProviderConfig", {
  enumerable: true,
  get: function () { return chunkOWZ6QT24_cjs.getProviderConfig; }
});
Object.defineProperty(exports, "getRegisteredProviders", {
  enumerable: true,
  get: function () { return chunkOWZ6QT24_cjs.getRegisteredProviders; }
});
Object.defineProperty(exports, "isProviderRegistered", {
  enumerable: true,
  get: function () { return chunkOWZ6QT24_cjs.isProviderRegistered; }
});
Object.defineProperty(exports, "isValidModelId", {
  enumerable: true,
  get: function () { return chunkOWZ6QT24_cjs.isValidModelId; }
});
Object.defineProperty(exports, "parseModelString", {
  enumerable: true,
  get: function () { return chunkOWZ6QT24_cjs.parseModelString; }
});
