// @mastra/memory/dist/processors/index.cjs
'use strict';

var chunkD6II7EP4_cjs = require('../chunk-D6II7EP4.cjs');



Object.defineProperty(exports, "OBSERVATIONAL_MEMORY_DEFAULTS", {
  enumerable: true,
  get: function () { return chunkD6II7EP4_cjs.OBSERVATIONAL_MEMORY_DEFAULTS; }
});
Object.defineProperty(exports, "OBSERVATION_CONTEXT_INSTRUCTIONS", {
  enumerable: true,
  get: function () { return chunkD6II7EP4_cjs.OBSERVATION_CONTEXT_INSTRUCTIONS; }
});
Object.defineProperty(exports, "OBSERVATION_CONTEXT_PROMPT", {
  enumerable: true,
  get: function () { return chunkD6II7EP4_cjs.OBSERVATION_CONTEXT_PROMPT; }
});
Object.defineProperty(exports, "OBSERVATION_CONTINUATION_HINT", {
  enumerable: true,
  get: function () { return chunkD6II7EP4_cjs.OBSERVATION_CONTINUATION_HINT; }
});
Object.defineProperty(exports, "OBSERVER_SYSTEM_PROMPT", {
  enumerable: true,
  get: function () { return chunkD6II7EP4_cjs.OBSERVER_SYSTEM_PROMPT; }
});
Object.defineProperty(exports, "ObservationalMemory", {
  enumerable: true,
  get: function () { return chunkD6II7EP4_cjs.ObservationalMemory; }
});
Object.defineProperty(exports, "TokenCounter", {
  enumerable: true,
  get: function () { return chunkD6II7EP4_cjs.TokenCounter; }
});
Object.defineProperty(exports, "buildObserverPrompt", {
  enumerable: true,
  get: function () { return chunkD6II7EP4_cjs.buildObserverPrompt; }
});
Object.defineProperty(exports, "buildObserverSystemPrompt", {
  enumerable: true,
  get: function () { return chunkD6II7EP4_cjs.buildObserverSystemPrompt; }
});
Object.defineProperty(exports, "extractCurrentTask", {
  enumerable: true,
  get: function () { return chunkD6II7EP4_cjs.extractCurrentTask; }
});
Object.defineProperty(exports, "formatMessagesForObserver", {
  enumerable: true,
  get: function () { return chunkD6II7EP4_cjs.formatMessagesForObserver; }
});
Object.defineProperty(exports, "hasCurrentTaskSection", {
  enumerable: true,
  get: function () { return chunkD6II7EP4_cjs.hasCurrentTaskSection; }
});
Object.defineProperty(exports, "optimizeObservationsForContext", {
  enumerable: true,
  get: function () { return chunkD6II7EP4_cjs.optimizeObservationsForContext; }
});
Object.defineProperty(exports, "parseObserverOutput", {
  enumerable: true,
  get: function () { return chunkD6II7EP4_cjs.parseObserverOutput; }
});
