// @mastra/server/dist/server/handlers/auth.cjs
'use strict';

var chunkCARPBU5W_cjs = require('../../chunk-CARPBU5W.cjs');



Object.defineProperty(exports, "AUTH_ROUTES", {
  enumerable: true,
  get: function () { return chunkCARPBU5W_cjs.AUTH_ROUTES; }
});
Object.defineProperty(exports, "GET_AUTH_CAPABILITIES_ROUTE", {
  enumerable: true,
  get: function () { return chunkCARPBU5W_cjs.GET_AUTH_CAPABILITIES_ROUTE; }
});
Object.defineProperty(exports, "GET_CURRENT_USER_ROUTE", {
  enumerable: true,
  get: function () { return chunkCARPBU5W_cjs.GET_CURRENT_USER_ROUTE; }
});
Object.defineProperty(exports, "GET_SSO_CALLBACK_ROUTE", {
  enumerable: true,
  get: function () { return chunkCARPBU5W_cjs.GET_SSO_CALLBACK_ROUTE; }
});
Object.defineProperty(exports, "GET_SSO_LOGIN_ROUTE", {
  enumerable: true,
  get: function () { return chunkCARPBU5W_cjs.GET_SSO_LOGIN_ROUTE; }
});
Object.defineProperty(exports, "POST_CREDENTIALS_SIGN_IN_ROUTE", {
  enumerable: true,
  get: function () { return chunkCARPBU5W_cjs.POST_CREDENTIALS_SIGN_IN_ROUTE; }
});
Object.defineProperty(exports, "POST_CREDENTIALS_SIGN_UP_ROUTE", {
  enumerable: true,
  get: function () { return chunkCARPBU5W_cjs.POST_CREDENTIALS_SIGN_UP_ROUTE; }
});
Object.defineProperty(exports, "POST_LOGOUT_ROUTE", {
  enumerable: true,
  get: function () { return chunkCARPBU5W_cjs.POST_LOGOUT_ROUTE; }
});
