// @mastra/server/dist/server/handlers/voice.cjs
'use strict';

var chunkBZO7NV5A_cjs = require('../../chunk-BZO7NV5A.cjs');



Object.defineProperty(exports, "GENERATE_SPEECH_DEPRECATED_ROUTE", {
  enumerable: true,
  get: function () { return chunkBZO7NV5A_cjs.GENERATE_SPEECH_DEPRECATED_ROUTE; }
});
Object.defineProperty(exports, "GENERATE_SPEECH_ROUTE", {
  enumerable: true,
  get: function () { return chunkBZO7NV5A_cjs.GENERATE_SPEECH_ROUTE; }
});
Object.defineProperty(exports, "GET_LISTENER_ROUTE", {
  enumerable: true,
  get: function () { return chunkBZO7NV5A_cjs.GET_LISTENER_ROUTE; }
});
Object.defineProperty(exports, "GET_SPEAKERS_DEPRECATED_ROUTE", {
  enumerable: true,
  get: function () { return chunkBZO7NV5A_cjs.GET_SPEAKERS_DEPRECATED_ROUTE; }
});
Object.defineProperty(exports, "GET_SPEAKERS_ROUTE", {
  enumerable: true,
  get: function () { return chunkBZO7NV5A_cjs.GET_SPEAKERS_ROUTE; }
});
Object.defineProperty(exports, "TRANSCRIBE_SPEECH_DEPRECATED_ROUTE", {
  enumerable: true,
  get: function () { return chunkBZO7NV5A_cjs.TRANSCRIBE_SPEECH_DEPRECATED_ROUTE; }
});
Object.defineProperty(exports, "TRANSCRIBE_SPEECH_ROUTE", {
  enumerable: true,
  get: function () { return chunkBZO7NV5A_cjs.TRANSCRIBE_SPEECH_ROUTE; }
});
