// @mastra/mcp/dist/index.cjs
'use strict';

var async_hooks = require('async_hooks');
var $RefParser = require('@apidevtools/json-schema-ref-parser');
var base = require('@mastra/core/base');
var error = require('@mastra/core/error');
var tools = require('@mastra/core/tools');
var utils = require('@mastra/core/utils');
var index_js = require('@modelcontextprotocol/sdk/client/index.js');
var sse_js = require('@modelcontextprotocol/sdk/client/sse.js');
var stdio_js = require('@modelcontextprotocol/sdk/client/stdio.js');
var streamableHttp_js = require('@modelcontextprotocol/sdk/client/streamableHttp.js');
var protocol_js = require('@modelcontextprotocol/sdk/shared/protocol.js');
var types_js = require('@modelcontextprotocol/sdk/types.js');
var exitHook = require('exit-hook');
var zod = require('zod');
var zodFromJsonSchema = require('zod-from-json-schema');
var zodFromJsonSchemaV3 = require('zod-from-json-schema-v3');
var equal = require('fast-deep-equal');
var uuid = require('uuid');
var crypto$1 = require('crypto');
var mcp = require('@mastra/core/mcp');
var requestContext = require('@mastra/core/request-context');
var index_js$1 = require('@modelcontextprotocol/sdk/server/index.js');
var sse_js$1 = require('@modelcontextprotocol/sdk/server/sse.js');
var stdio_js$1 = require('@modelcontextprotocol/sdk/server/stdio.js');
var streamableHttp_js$1 = require('@modelcontextprotocol/sdk/server/streamableHttp.js');
var auth_js = require('@modelcontextprotocol/sdk/client/auth.js');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var $RefParser__default = /*#__PURE__*/_interopDefault($RefParser);
var equal__default = /*#__PURE__*/_interopDefault(equal);

// src/client/client.ts

// src/client/actions/elicitation.ts
var ElicitationClientActions = class {
  client;
  logger;
  /**
   * @internal
   */
  constructor({ client, logger }) {
    this.client = client;
    this.logger = logger;
  }
  /**
   * Sets a handler function for processing elicitation requests from the server.
   *
   * The handler is called when the server needs to collect user input during tool execution.
   * The handler must return a response with action ('accept', 'decline', or 'cancel') and
   * optional content matching the requested schema.
   *
   * @param handler - Callback function to handle elicitation requests
   *
   * @example
   * ```typescript
   * client.elicitation.onRequest(async (request) => {
   *   console.log('Server message:', request.message);
   *   console.log('Requested schema:', request.requestedSchema);
   *
   *   // Collect user input (e.g., via CLI prompt or UI form)
   *   const userInput = await collectUserInput(request.requestedSchema);
   *
   *   return {
   *     action: 'accept',
   *     content: userInput
   *   };
   * });
   * ```
   *
   * @example
   * ```typescript
   * // Declining an elicitation request
   * client.elicitation.onRequest(async (request) => {
   *   if (!shouldAcceptRequest(request)) {
   *     return { action: 'decline' };
   *   }
   *
   *   const input = await getInput();
   *   return { action: 'accept', content: input };
   * });
   * ```
   */
  onRequest(handler) {
    this.client.setElicitationRequestHandler(handler);
  }
};

// src/client/actions/progress.ts
var ProgressClientActions = class {
  client;
  logger;
  constructor({ client, logger }) {
    this.client = client;
    this.logger = logger;
  }
  /**
   * Set a notification handler for progress updates.
   * @param handler The callback function to handle progress notifications.
   */
  onUpdate(handler) {
    this.client.setProgressNotificationHandler(handler);
  }
};
var PromptClientActions = class {
  client;
  logger;
  /**
   * @internal
   */
  constructor({ client, logger }) {
    this.client = client;
    this.logger = logger;
  }
  /**
   * Retrieves all available prompts from the connected MCP server.
   *
   * Returns an empty array if the server doesn't support prompts (MethodNotFound error).
   *
   * @returns Promise resolving to array of prompts with their metadata
   * @throws {Error} If fetching prompts fails (excluding MethodNotFound)
   *
   * @example
   * ```typescript
   * const prompts = await client.prompts.list();
   * prompts.forEach(prompt => {
   *   console.log(`${prompt.name} (v${prompt.version}): ${prompt.description}`);
   * });
   * ```
   */
  async list() {
    try {
      const response = await this.client.listPrompts();
      if (response && response.prompts && Array.isArray(response.prompts)) {
        return response.prompts.map((prompt) => ({ ...prompt, version: prompt.version || "" }));
      } else {
        this.logger.warn(`Prompts response from server ${this.client.name} did not have expected structure.`, {
          response
        });
        return [];
      }
    } catch (e) {
      if (e.code === types_js.ErrorCode.MethodNotFound) {
        return [];
      }
      this.logger.error(`Error getting prompts from server ${this.client.name}`, {
        error: e instanceof Error ? e.message : String(e)
      });
      throw new Error(
        `Failed to fetch prompts from server ${this.client.name}: ${e instanceof Error ? e.stack || e.message : String(e)}`
      );
    }
  }
  /**
   * Retrieves a specific prompt with its messages from the MCP server.
   *
   * Prompts can accept arguments to parameterize the template. The returned messages
   * can be used directly in AI chat completions.
   *
   * @param params - Parameters for the prompt request
   * @param params.name - Name of the prompt to retrieve
   * @param params.args - Optional arguments to populate the prompt template
   * @param params.version - Optional specific version of the prompt to retrieve
   * @returns Promise resolving to the prompt result with messages
   * @throws {Error} If fetching the prompt fails or prompt not found
   *
   * @example
   * ```typescript
   * const prompt = await client.prompts.get({
   *   name: 'code-review',
   *   args: {
   *     language: 'typescript',
   *     code: 'const x = 1;'
   *   },
   *   version: '1.0'
   * });
   *
   * // Use prompt messages in AI completion
   * console.log(prompt.messages);
   * ```
   */
  async get({
    name,
    args,
    version
  }) {
    return this.client.getPrompt({ name, args, version });
  }
  /**
   * Sets a notification handler for when the list of available prompts changes.
   *
   * The handler is called when prompts are added, removed, or modified on the server.
   *
   * @param handler - Callback function invoked when the prompt list changes
   *
   * @example
   * ```typescript
   * await client.prompts.onListChanged(async () => {
   *   console.log('Prompt list changed, re-fetching...');
   *   const prompts = await client.prompts.list();
   *   console.log('Available prompts:', prompts.map(p => p.name));
   * });
   * ```
   */
  async onListChanged(handler) {
    this.client.setPromptListChangedNotificationHandler(handler);
  }
};
var ResourceClientActions = class {
  client;
  logger;
  /**
   * @internal
   */
  constructor({ client, logger }) {
    this.client = client;
    this.logger = logger;
  }
  /**
   * Retrieves all available resources from the connected MCP server.
   *
   * Returns an empty array if the server doesn't support resources (MethodNotFound error).
   *
   * @returns Promise resolving to array of resources
   * @throws {Error} If fetching resources fails (excluding MethodNotFound)
   *
   * @example
   * ```typescript
   * const resources = await client.resources.list();
   * resources.forEach(resource => {
   *   console.log(`${resource.name}: ${resource.uri}`);
   * });
   * ```
   */
  async list() {
    try {
      const response = await this.client.listResources();
      if (response && response.resources && Array.isArray(response.resources)) {
        return response.resources;
      } else {
        this.logger.warn(`Resources response from server ${this.client.name} did not have expected structure.`, {
          response
        });
        return [];
      }
    } catch (e) {
      if (e.code === types_js.ErrorCode.MethodNotFound) {
        return [];
      }
      this.logger.error(`Error getting resources from server ${this.client.name}`, {
        error: e instanceof Error ? e.message : String(e)
      });
      throw new Error(
        `Failed to fetch resources from server ${this.client.name}: ${e instanceof Error ? e.stack || e.message : String(e)}`
      );
    }
  }
  /**
   * Retrieves all available resource templates from the connected MCP server.
   *
   * Resource templates are URI templates (RFC 6570) that describe dynamic resources.
   * Returns an empty array if the server doesn't support resource templates.
   *
   * @returns Promise resolving to array of resource templates
   * @throws {Error} If fetching resource templates fails (excluding MethodNotFound)
   *
   * @example
   * ```typescript
   * const templates = await client.resources.templates();
   * templates.forEach(template => {
   *   console.log(`${template.name}: ${template.uriTemplate}`);
   * });
   * ```
   */
  async templates() {
    try {
      const response = await this.client.listResourceTemplates();
      if (response && response.resourceTemplates && Array.isArray(response.resourceTemplates)) {
        return response.resourceTemplates;
      } else {
        this.logger.warn(
          `Resource templates response from server ${this.client.name} did not have expected structure.`,
          { response }
        );
        return [];
      }
    } catch (e) {
      if (e.code === types_js.ErrorCode.MethodNotFound) {
        return [];
      }
      this.logger.error(`Error getting resource templates from server ${this.client.name}`, {
        error: e instanceof Error ? e.message : String(e)
      });
      throw new Error(
        `Failed to fetch resource templates from server ${this.client.name}: ${e instanceof Error ? e.stack || e.message : String(e)}`
      );
    }
  }
  /**
   * Reads the content of a specific resource from the MCP server.
   *
   * @param uri - URI of the resource to read (e.g., 'file://path/to/file.txt')
   * @returns Promise resolving to the resource content
   * @throws {Error} If reading the resource fails or resource not found
   *
   * @example
   * ```typescript
   * const result = await client.resources.read('file://data/config.json');
   * console.log(result.contents[0].text); // Resource text content
   * ```
   */
  async read(uri) {
    return this.client.readResource(uri);
  }
  /**
   * Subscribes to updates for a specific resource.
   *
   * After subscribing, you'll receive notifications via the `onUpdated` handler
   * when the resource content changes.
   *
   * @param uri - URI of the resource to subscribe to
   * @returns Promise resolving when subscription is established
   * @throws {Error} If subscription fails
   *
   * @example
   * ```typescript
   * await client.resources.subscribe('file://data/config.json');
   * ```
   */
  async subscribe(uri) {
    return this.client.subscribeResource(uri);
  }
  /**
   * Unsubscribes from updates for a specific resource.
   *
   * Stops receiving notifications for this resource URI.
   *
   * @param uri - URI of the resource to unsubscribe from
   * @returns Promise resolving when unsubscription is complete
   * @throws {Error} If unsubscription fails
   *
   * @example
   * ```typescript
   * await client.resources.unsubscribe('file://data/config.json');
   * ```
   */
  async unsubscribe(uri) {
    return this.client.unsubscribeResource(uri);
  }
  /**
   * Sets a notification handler for when subscribed resources are updated.
   *
   * The handler is called whenever the server sends a resource update notification
   * for any resource you've subscribed to.
   *
   * @param handler - Callback function receiving the updated resource URI
   *
   * @example
   * ```typescript
   * await client.resources.onUpdated(async (params) => {
   *   console.log(`Resource updated: ${params.uri}`);
   *   // Re-fetch the resource
   *   const content = await client.resources.read(params.uri);
   *   console.log('New content:', content);
   * });
   * ```
   */
  async onUpdated(handler) {
    this.client.setResourceUpdatedNotificationHandler(handler);
  }
  /**
   * Sets a notification handler for when the list of available resources changes.
   *
   * The handler is called when resources are added or removed from the server.
   *
   * @param handler - Callback function invoked when the resource list changes
   *
   * @example
   * ```typescript
   * await client.resources.onListChanged(async () => {
   *   console.log('Resource list changed, re-fetching...');
   *   const resources = await client.resources.list();
   *   console.log('Updated resource count:', resources.length);
   * });
   * ```
   */
  async onListChanged(handler) {
    this.client.setResourceListChangedNotificationHandler(handler);
  }
};

// src/client/client.ts
var DEFAULT_SERVER_CONNECT_TIMEOUT_MSEC = 3e3;
var SSE_FALLBACK_STATUS_CODES = [400, 404, 405];
function convertLogLevelToLoggerMethod(level) {
  switch (level) {
    case "debug":
      return "debug";
    case "info":
    case "notice":
      return "info";
    case "warning":
      return "warn";
    case "error":
    case "critical":
    case "alert":
    case "emergency":
      return "error";
    default:
      return "info";
  }
}
var InternalMastraMCPClient = class extends base.MastraBase {
  name;
  client;
  timeout;
  logHandler;
  enableServerLogs;
  enableProgressTracking;
  serverConfig;
  transport;
  operationContextStore = new async_hooks.AsyncLocalStorage();
  exitHookUnsubscribe;
  sigTermHandler;
  _roots;
  /** Provides access to resource operations (list, read, subscribe, etc.) */
  resources;
  /** Provides access to prompt operations (list, get, notifications) */
  prompts;
  /** Provides access to elicitation operations (request handling) */
  elicitation;
  /** Provides access to progress operations (notifications) */
  progress;
  /**
   * @internal
   */
  constructor({
    name,
    version = "1.0.0",
    server,
    capabilities = {},
    timeout = protocol_js.DEFAULT_REQUEST_TIMEOUT_MSEC
  }) {
    super({ name: "MastraMCPClient" });
    this.name = name;
    this.timeout = timeout;
    this.logHandler = server.logger;
    this.enableServerLogs = server.enableServerLogs ?? true;
    this.serverConfig = server;
    this.enableProgressTracking = !!server.enableProgressTracking;
    this._roots = server.roots ?? [];
    const hasRoots = this._roots.length > 0 || !!capabilities.roots;
    const clientCapabilities = {
      ...capabilities,
      elicitation: {},
      // Auto-enable roots capability if roots are provided
      ...hasRoots ? { roots: { listChanged: true, ...capabilities.roots ?? {} } } : {}
    };
    this.client = new index_js.Client(
      {
        name,
        version
      },
      {
        capabilities: clientCapabilities
      }
    );
    this.setupLogging();
    if (hasRoots) {
      this.setupRootsHandler();
    }
    this.resources = new ResourceClientActions({ client: this, logger: this.logger });
    this.prompts = new PromptClientActions({ client: this, logger: this.logger });
    this.elicitation = new ElicitationClientActions({ client: this, logger: this.logger });
    this.progress = new ProgressClientActions({ client: this, logger: this.logger });
  }
  /**
   * Log a message at the specified level
   * @param level Log level
   * @param message Log message
   * @param details Optional additional details
   */
  log(level, message, details) {
    const loggerMethod = convertLogLevelToLoggerMethod(level);
    const msg = `[${this.name}] ${message}`;
    this.logger[loggerMethod](msg, details);
    if (this.logHandler) {
      this.logHandler({
        level,
        message: msg,
        timestamp: /* @__PURE__ */ new Date(),
        serverName: this.name,
        details,
        requestContext: this.operationContextStore.getStore() ?? null
      });
    }
  }
  setupLogging() {
    if (this.enableServerLogs) {
      this.client.setNotificationHandler(
        zod.z.object({
          method: zod.z.literal("notifications/message"),
          params: zod.z.object({
            level: zod.z.string()
          }).passthrough()
        }),
        (notification) => {
          const { level, ...params } = notification.params;
          this.log(level, "[MCP SERVER LOG]", params);
        }
      );
    }
  }
  /**
   * Set up handler for roots/list requests from the server.
   *
   * Per MCP spec (https://modelcontextprotocol.io/specification/2025-11-25/client/roots):
   * When a server sends a roots/list request, the client responds with the configured roots.
   */
  setupRootsHandler() {
    this.log("debug", "Setting up roots/list request handler");
    this.client.setRequestHandler(types_js.ListRootsRequestSchema, async () => {
      this.log("debug", `Responding to roots/list request with ${this._roots.length} roots`);
      return { roots: this._roots };
    });
  }
  /**
   * Get the currently configured roots.
   *
   * @returns Array of configured filesystem roots
   */
  get roots() {
    return [...this._roots];
  }
  /**
   * Update the list of filesystem roots and notify the server.
   *
   * Per MCP spec, when roots change, the client sends a `notifications/roots/list_changed`
   * notification to inform the server that it should re-fetch the roots list.
   *
   * @param roots - New list of filesystem roots
   *
   * @example
   * ```typescript
   * await client.setRoots([
   *   { uri: 'file:///home/user/projects', name: 'Projects' },
   *   { uri: 'file:///tmp', name: 'Temp' }
   * ]);
   * ```
   */
  async setRoots(roots) {
    this.log("debug", `Updating roots to ${roots.length} entries`);
    this._roots = [...roots];
    await this.sendRootsListChanged();
  }
  /**
   * Send a roots/list_changed notification to the server.
   *
   * Per MCP spec, clients that support `listChanged` MUST send this notification
   * when the list of roots changes. The server will then call roots/list to get
   * the updated list.
   */
  async sendRootsListChanged() {
    if (!this.transport) {
      this.log("debug", "Cannot send roots/list_changed: not connected");
      return;
    }
    this.log("debug", "Sending notifications/roots/list_changed");
    await this.client.notification({ method: "notifications/roots/list_changed" });
  }
  async connectStdio(command) {
    this.log("debug", `Using Stdio transport for command: ${command}`);
    try {
      this.transport = new stdio_js.StdioClientTransport({
        command,
        args: this.serverConfig.args,
        env: { ...stdio_js.getDefaultEnvironment(), ...this.serverConfig.env || {} }
      });
      await this.client.connect(this.transport, { timeout: this.serverConfig.timeout ?? this.timeout });
      this.log("debug", `Successfully connected to MCP server via Stdio`);
    } catch (e) {
      this.log("error", e instanceof Error ? e.stack || e.message : JSON.stringify(e));
      throw e;
    }
  }
  async connectHttp(url) {
    const { requestInit, eventSourceInit, authProvider, connectTimeout, fetch: userFetch } = this.serverConfig;
    const fetch2 = userFetch ? (url2, init) => userFetch(url2, init, this.operationContextStore.getStore() ?? null) : void 0;
    this.log("debug", `Attempting to connect to URL: ${url}`);
    let shouldTrySSE = url.pathname.endsWith(`/sse`);
    if (!shouldTrySSE) {
      try {
        this.log("debug", "Trying Streamable HTTP transport...");
        const streamableTransport = new streamableHttp_js.StreamableHTTPClientTransport(url, {
          requestInit,
          reconnectionOptions: this.serverConfig.reconnectionOptions,
          authProvider,
          fetch: fetch2
        });
        await this.client.connect(streamableTransport, {
          timeout: connectTimeout ?? DEFAULT_SERVER_CONNECT_TIMEOUT_MSEC
        });
        this.transport = streamableTransport;
        this.log("debug", "Successfully connected using Streamable HTTP transport.");
      } catch (error) {
        this.log("debug", `Streamable HTTP transport failed: ${error}`);
        const status = error?.code;
        if (status !== void 0 && !SSE_FALLBACK_STATUS_CODES.includes(status)) {
          throw error;
        }
        shouldTrySSE = true;
      }
    }
    if (shouldTrySSE) {
      this.log("debug", "Falling back to deprecated HTTP+SSE transport...");
      try {
        const sseEventSourceInit = fetch2 ? { ...eventSourceInit, fetch: fetch2 } : eventSourceInit;
        const sseTransport = new sse_js.SSEClientTransport(url, {
          requestInit,
          eventSourceInit: sseEventSourceInit,
          authProvider,
          fetch: fetch2
        });
        await this.client.connect(sseTransport, { timeout: this.serverConfig.timeout ?? this.timeout });
        this.transport = sseTransport;
        this.log("debug", "Successfully connected using deprecated HTTP+SSE transport.");
      } catch (sseError) {
        this.log(
          "error",
          `Failed to connect with SSE transport after failing to connect to Streamable HTTP transport first. SSE error: ${sseError}`
        );
        throw new Error("Could not connect to server with any available HTTP transport");
      }
    }
  }
  isConnected = null;
  /**
   * Connects to the MCP server using the configured transport.
   *
   * Automatically detects transport type based on configuration (stdio vs HTTP).
   * Safe to call multiple times - returns existing connection if already connected.
   *
   * @returns Promise resolving to true when connected
   * @throws {MastraError} If connection fails
   *
   * @internal
   */
  async connect() {
    if (this.isConnected) {
      return this.isConnected;
    }
    this.isConnected = new Promise(async (resolve, reject) => {
      try {
        const { command, url } = this.serverConfig;
        if (command) {
          await this.connectStdio(command);
        } else if (url) {
          await this.connectHttp(url);
        } else {
          throw new Error("Server configuration must include either a command or a url.");
        }
        resolve(true);
        const originalOnClose = this.client.onclose;
        this.client.onclose = () => {
          this.log("debug", `MCP server connection closed`);
          this.isConnected = null;
          if (typeof originalOnClose === "function") {
            originalOnClose();
          }
        };
      } catch (e) {
        this.isConnected = null;
        reject(e);
      }
    });
    if (!this.exitHookUnsubscribe) {
      this.exitHookUnsubscribe = exitHook.asyncExitHook(
        async () => {
          this.log("debug", `Disconnecting MCP server during exit`);
          await this.disconnect();
        },
        { wait: 5e3 }
      );
    }
    if (!this.sigTermHandler) {
      this.sigTermHandler = () => exitHook.gracefulExit();
      process.on("SIGTERM", this.sigTermHandler);
    }
    this.log("debug", `Successfully connected to MCP server`);
    return this.isConnected;
  }
  /**
   * Gets the current session ID if using Streamable HTTP transport.
   *
   * Returns undefined if not connected or not using Streamable HTTP transport.
   *
   * @returns Session ID string or undefined
   *
   * @internal
   */
  get sessionId() {
    if (this.transport instanceof streamableHttp_js.StreamableHTTPClientTransport) {
      return this.transport.sessionId;
    }
    return void 0;
  }
  async disconnect() {
    if (!this.transport) {
      this.log("debug", "Disconnect called but no transport was connected.");
      return;
    }
    this.log("debug", `Disconnecting from MCP server`);
    try {
      await this.transport.close();
      this.log("debug", "Successfully disconnected from MCP server");
    } catch (e) {
      this.log("error", "Error during MCP server disconnect", {
        error: e instanceof Error ? e.stack : JSON.stringify(e, null, 2)
      });
      throw e;
    } finally {
      this.transport = void 0;
      this.isConnected = null;
      if (this.exitHookUnsubscribe) {
        this.exitHookUnsubscribe();
        this.exitHookUnsubscribe = void 0;
      }
      if (this.sigTermHandler) {
        process.off("SIGTERM", this.sigTermHandler);
        this.sigTermHandler = void 0;
      }
    }
  }
  /**
   * Checks if an error indicates a session invalidation that requires reconnection.
   *
   * Common session-related errors include:
   * - "No valid session ID provided" (HTTP 400)
   * - "Server not initialized" (HTTP 400)
   * - "Not connected" (protocol state error)
   * - Connection refused errors
   *
   * @param error - The error to check
   * @returns true if the error indicates a session problem requiring reconnection
   *
   * @internal
   */
  isSessionError(error) {
    if (!(error instanceof Error)) {
      return false;
    }
    const errorMessage = error.message.toLowerCase();
    return errorMessage.includes("no valid session") || errorMessage.includes("session") || errorMessage.includes("server not initialized") || errorMessage.includes("not connected") || errorMessage.includes("http 400") || errorMessage.includes("http 401") || errorMessage.includes("http 403") || errorMessage.includes("econnrefused") || errorMessage.includes("fetch failed") || errorMessage.includes("connection refused") || errorMessage.includes("sse stream disconnected") || errorMessage.includes("typeerror: terminated");
  }
  /**
   * Forces a reconnection to the MCP server by disconnecting and reconnecting.
   *
   * This is useful when the session becomes invalid (e.g., after server restart)
   * and the client needs to establish a fresh connection.
   *
   * @returns Promise resolving when reconnection is complete
   * @throws {Error} If reconnection fails
   *
   * @internal
   */
  async forceReconnect() {
    this.log("debug", "Forcing reconnection to MCP server...");
    try {
      if (this.transport) {
        await this.transport.close();
      }
    } catch (e) {
      this.log("debug", "Error during force disconnect (ignored)", {
        error: e instanceof Error ? e.message : String(e)
      });
    }
    this.transport = void 0;
    this.isConnected = null;
    await this.connect();
    this.log("debug", "Successfully reconnected to MCP server");
  }
  async listResources() {
    this.log("debug", `Requesting resources from MCP server`);
    return await this.client.request({ method: "resources/list" }, types_js.ListResourcesResultSchema, {
      timeout: this.timeout
    });
  }
  async readResource(uri) {
    this.log("debug", `Reading resource from MCP server: ${uri}`);
    return await this.client.request({ method: "resources/read", params: { uri } }, types_js.ReadResourceResultSchema, {
      timeout: this.timeout
    });
  }
  async subscribeResource(uri) {
    this.log("debug", `Subscribing to resource on MCP server: ${uri}`);
    return await this.client.request({ method: "resources/subscribe", params: { uri } }, zod.z.object({}), {
      timeout: this.timeout
    });
  }
  async unsubscribeResource(uri) {
    this.log("debug", `Unsubscribing from resource on MCP server: ${uri}`);
    return await this.client.request({ method: "resources/unsubscribe", params: { uri } }, zod.z.object({}), {
      timeout: this.timeout
    });
  }
  async listResourceTemplates() {
    this.log("debug", `Requesting resource templates from MCP server`);
    return await this.client.request({ method: "resources/templates/list" }, types_js.ListResourceTemplatesResultSchema, {
      timeout: this.timeout
    });
  }
  /**
   * Fetch the list of available prompts from the MCP server.
   */
  async listPrompts() {
    this.log("debug", `Requesting prompts from MCP server`);
    return await this.client.request({ method: "prompts/list" }, types_js.ListPromptsResultSchema, {
      timeout: this.timeout
    });
  }
  /**
   * Get a prompt and its dynamic messages from the server.
   * @param name The prompt name
   * @param args Arguments for the prompt
   * @param version (optional) The prompt version to retrieve
   */
  async getPrompt({
    name,
    args,
    version
  }) {
    this.log("debug", `Requesting prompt from MCP server: ${name}`);
    return await this.client.request(
      { method: "prompts/get", params: { name, arguments: args, version } },
      types_js.GetPromptResultSchema,
      { timeout: this.timeout }
    );
  }
  /**
   * Register a handler to be called when the prompt list changes on the server.
   * Use this to refresh cached prompt lists in the client/UI if needed.
   */
  setPromptListChangedNotificationHandler(handler) {
    this.log("debug", "Setting prompt list changed notification handler");
    this.client.setNotificationHandler(types_js.PromptListChangedNotificationSchema, () => {
      handler();
    });
  }
  setResourceUpdatedNotificationHandler(handler) {
    this.log("debug", "Setting resource updated notification handler");
    this.client.setNotificationHandler(types_js.ResourceUpdatedNotificationSchema, (notification) => {
      handler(notification.params);
    });
  }
  setResourceListChangedNotificationHandler(handler) {
    this.log("debug", "Setting resource list changed notification handler");
    this.client.setNotificationHandler(types_js.ResourceListChangedNotificationSchema, () => {
      handler();
    });
  }
  setElicitationRequestHandler(handler) {
    this.log("debug", "Setting elicitation request handler");
    this.client.setRequestHandler(types_js.ElicitRequestSchema, async (request) => {
      this.log("debug", `Received elicitation request: ${request.params.message}`);
      return handler(request.params);
    });
  }
  setProgressNotificationHandler(handler) {
    this.log("debug", "Setting progress notification handler");
    this.client.setNotificationHandler(types_js.ProgressNotificationSchema, (notification) => {
      handler(notification.params);
    });
  }
  async convertInputSchema(inputSchema) {
    if (utils.isZodType(inputSchema)) {
      return inputSchema;
    }
    try {
      await $RefParser__default.default.dereference(inputSchema);
      const jsonSchemaToConvert = "jsonSchema" in inputSchema ? inputSchema.jsonSchema : inputSchema;
      if ("toJSONSchema" in zod.z) {
        return zodFromJsonSchema.convertJsonSchemaToZod(jsonSchemaToConvert);
      } else {
        return zodFromJsonSchemaV3.convertJsonSchemaToZod(jsonSchemaToConvert);
      }
    } catch (error$1) {
      let errorDetails;
      if (error$1 instanceof Error) {
        errorDetails = error$1.stack;
      } else {
        try {
          errorDetails = JSON.stringify(error$1);
        } catch {
          errorDetails = String(error$1);
        }
      }
      this.log("error", "Failed to convert JSON schema to Zod schema using zodFromJsonSchema", {
        error: errorDetails,
        originalJsonSchema: inputSchema
      });
      throw new error.MastraError({
        id: "MCP_TOOL_INPUT_SCHEMA_CONVERSION_FAILED",
        domain: error.ErrorDomain.MCP,
        category: error.ErrorCategory.USER,
        details: { error: errorDetails ?? "Unknown error" }
      });
    }
  }
  /**
   * Recursively applies `.passthrough()` to all ZodObject schemas so that
   * unknown keys returned by an MCP server are preserved instead of being
   * silently stripped by Zod's default "strip" mode.
   */
  applyPassthrough(schema) {
    if (schema instanceof zod.z.ZodObject) {
      const shape = schema.shape;
      const newShape = {};
      for (const key of Object.keys(shape)) {
        newShape[key] = this.applyPassthrough(shape[key]);
      }
      return zod.z.object(newShape).passthrough();
    }
    if (schema instanceof zod.z.ZodArray) {
      return zod.z.array(this.applyPassthrough(schema.element));
    }
    if (schema instanceof zod.z.ZodOptional) {
      return this.applyPassthrough(schema.unwrap()).optional();
    }
    if (schema instanceof zod.z.ZodNullable) {
      return this.applyPassthrough(schema.unwrap()).nullable();
    }
    return schema;
  }
  async convertOutputSchema(outputSchema) {
    if (!outputSchema) return;
    if (utils.isZodType(outputSchema)) {
      return outputSchema;
    }
    try {
      await $RefParser__default.default.dereference(outputSchema);
      const jsonSchemaToConvert = "jsonSchema" in outputSchema ? outputSchema.jsonSchema : outputSchema;
      let zodSchema;
      if ("toJSONSchema" in zod.z) {
        zodSchema = zodFromJsonSchema.convertJsonSchemaToZod(jsonSchemaToConvert);
      } else {
        zodSchema = zodFromJsonSchemaV3.convertJsonSchemaToZod(jsonSchemaToConvert);
      }
      return this.applyPassthrough(zodSchema);
    } catch (error$1) {
      let errorDetails;
      if (error$1 instanceof Error) {
        errorDetails = error$1.stack;
      } else {
        try {
          errorDetails = JSON.stringify(error$1);
        } catch {
          errorDetails = String(error$1);
        }
      }
      this.log("error", "Failed to convert JSON schema to Zod schema using zodFromJsonSchema", {
        error: errorDetails,
        originalJsonSchema: outputSchema
      });
      throw new error.MastraError({
        id: "MCP_TOOL_OUTPUT_SCHEMA_CONVERSION_FAILED",
        domain: error.ErrorDomain.MCP,
        category: error.ErrorCategory.USER,
        details: { error: errorDetails ?? "Unknown error" }
      });
    }
  }
  async tools() {
    this.log("debug", `Requesting tools from MCP server`);
    const { tools: tools$1 } = await this.client.listTools({}, { timeout: this.timeout });
    const toolsRes = {};
    for (const tool of tools$1) {
      this.log("debug", `Processing tool: ${tool.name}`);
      try {
        const mastraTool = tools.createTool({
          id: `${this.name}_${tool.name}`,
          description: tool.description || "",
          inputSchema: await this.convertInputSchema(tool.inputSchema),
          outputSchema: await this.convertOutputSchema(tool.outputSchema),
          execute: async (input, context) => {
            const operationContext = context?.requestContext ?? null;
            return this.operationContextStore.run(operationContext, async () => {
              const executeToolCall = async () => {
                this.log("debug", `Executing tool: ${tool.name}`, { toolArgs: input, runId: context?.runId });
                const res = await this.client.callTool(
                  {
                    name: tool.name,
                    arguments: input,
                    // Use runId as progress token if available, otherwise generate a random UUID
                    ...this.enableProgressTracking ? { _meta: { progressToken: context?.runId || crypto.randomUUID() } } : {}
                  },
                  types_js.CallToolResultSchema,
                  {
                    timeout: this.timeout
                  }
                );
                this.log("debug", `Tool executed successfully: ${tool.name}`);
                if (res.structuredContent !== void 0) {
                  return res.structuredContent;
                }
                if (tool.outputSchema && !res.isError) {
                  const content = res.content;
                  if (content && content.length === 1 && content[0].type === "text" && content[0].text !== void 0) {
                    try {
                      return JSON.parse(content[0].text);
                    } catch {
                      return content[0].text;
                    }
                  }
                }
                return res;
              };
              try {
                return await executeToolCall();
              } catch (e) {
                if (this.isSessionError(e)) {
                  this.log("debug", `Session error detected for tool ${tool.name}, attempting reconnection...`, {
                    error: e instanceof Error ? e.message : String(e)
                  });
                  try {
                    await this.forceReconnect();
                    this.log("debug", `Retrying tool ${tool.name} after reconnection...`);
                    return await executeToolCall();
                  } catch (reconnectError) {
                    this.log("error", `Reconnection or retry failed for tool ${tool.name}`, {
                      originalError: e instanceof Error ? e.message : String(e),
                      reconnectError: reconnectError instanceof Error ? reconnectError.stack : String(reconnectError),
                      toolArgs: input
                    });
                    throw e;
                  }
                }
                this.log("error", `Error calling tool: ${tool.name}`, {
                  error: e instanceof Error ? e.stack : JSON.stringify(e, null, 2),
                  toolArgs: input
                });
                throw e;
              }
            });
          }
        });
        if (tool.name) {
          toolsRes[tool.name] = mastraTool;
        }
      } catch (toolCreationError) {
        this.log("error", `Failed to create Mastra tool wrapper for MCP tool: ${tool.name}`, {
          error: toolCreationError instanceof Error ? toolCreationError.stack : String(toolCreationError),
          mcpToolDefinition: tool
        });
      }
    }
    return toolsRes;
  }
};
var mcpClientInstances = /* @__PURE__ */ new Map();
var MCPClient = class extends base.MastraBase {
  serverConfigs = {};
  id;
  defaultTimeout;
  mcpClientsById = /* @__PURE__ */ new Map();
  disconnectPromise = null;
  /**
   * Creates a new MCPClient instance for managing MCP server connections.
   *
   * The client automatically manages connection lifecycle and prevents memory leaks by
   * caching instances with identical configurations.
   *
   * @param args - Configuration options
   * @param args.id - Optional unique identifier to allow multiple instances with same config
   * @param args.servers - Map of server names to server configurations
   * @param args.timeout - Optional global timeout in milliseconds (default: 60000)
   *
   * @throws {Error} If multiple instances with identical config are created without an ID
   *
   * @example
   * ```typescript
   * const mcp = new MCPClient({
   *   servers: {
   *     weatherServer: {
   *       url: new URL('http://localhost:8080/sse'),
   *       requestInit: {
   *         headers: { Authorization: 'Bearer token' }
   *       }
   *     }
   *   },
   *   timeout: 30000
   * });
   * ```
   */
  constructor(args) {
    super({ name: "MCPClient" });
    this.defaultTimeout = args.timeout ?? protocol_js.DEFAULT_REQUEST_TIMEOUT_MSEC;
    this.serverConfigs = args.servers;
    this.id = args.id ?? this.makeId();
    if (args.id) {
      this.id = args.id;
      const cached = mcpClientInstances.get(this.id);
      if (cached && !equal__default.default(cached.serverConfigs, args.servers)) {
        const existingInstance2 = mcpClientInstances.get(this.id);
        if (existingInstance2) {
          void existingInstance2.disconnect();
          mcpClientInstances.delete(this.id);
        }
      }
    } else {
      this.id = this.makeId();
    }
    const existingInstance = mcpClientInstances.get(this.id);
    if (existingInstance) {
      if (!args.id) {
        throw new Error(`MCPClient was initialized multiple times with the same configuration options.

This error is intended to prevent memory leaks.

To fix this you have three different options:
1. If you need multiple MCPClient class instances with identical server configurations, set an id when configuring: new MCPClient({ id: "my-unique-id" })
2. Call "await client.disconnect()" after you're done using the client and before you recreate another instance with the same options. If the identical MCPClient instance is already closed at the time of re-creating it, you will not see this error.
3. If you only need one instance of MCPClient in your app, refactor your code so it's only created one time (ex. move it out of a loop into a higher scope code block)
`);
      }
      return existingInstance;
    }
    mcpClientInstances.set(this.id, this);
    this.addToInstanceCache();
    return this;
  }
  /**
   * Provides access to progress-related operations for tracking long-running operations.
   *
   * Progress tracking allows MCP servers to send updates about the status of ongoing operations,
   * providing real-time feedback to users about task completion and current state.
   *
   * @example
   * ```typescript
   * // Set up handler for progress updates from a server
   * await mcp.progress.onUpdate('serverName', (params) => {
   *   console.log(`Progress: ${params.progress}%`);
   *   console.log(`Status: ${params.message}`);
   *
   *   if (params.total) {
   *     console.log(`Completed ${params.progress} of ${params.total} items`);
   *   }
   * });
   * ```
   */
  get progress() {
    this.addToInstanceCache();
    return {
      onUpdate: async (serverName, handler) => {
        try {
          const internalClient = await this.getConnectedClientForServer(serverName);
          return internalClient.progress.onUpdate(handler);
        } catch (err) {
          throw new error.MastraError(
            {
              id: "MCP_CLIENT_ON_UPDATE_PROGRESS_FAILED",
              domain: error.ErrorDomain.MCP,
              category: error.ErrorCategory.THIRD_PARTY,
              details: {
                serverName
              }
            },
            err
          );
        }
      }
    };
  }
  /**
   * Provides access to elicitation-related operations for interactive user input collection.
   *
   * Elicitation allows MCP servers to request structured information from users during tool execution.
   *
   * @example
   * ```typescript
   * // Set up handler for elicitation requests from a server
   * await mcp.elicitation.onRequest('serverName', async (request) => {
   *   console.log(`Server requests: ${request.message}`);
   *   console.log('Schema:', request.requestedSchema);
   *
   *   // Collect user input and return response
   *   return {
   *     action: 'accept',
   *     content: { name: 'John Doe', email: 'john@example.com' }
   *   };
   * });
   * ```
   */
  get elicitation() {
    this.addToInstanceCache();
    return {
      /**
       * Sets up a handler function for elicitation requests from a specific server.
       *
       * The handler receives requests for user input and must return a response with
       * action ('accept', 'decline', or 'cancel') and optional content.
       *
       * @param serverName - Name of the server to handle elicitation requests for
       * @param handler - Function to handle elicitation requests
       * @throws {MastraError} If setting up the handler fails
       *
       * @example
       * ```typescript
       * await mcp.elicitation.onRequest('weatherServer', async (request) => {
       *   // Prompt user for input
       *   const userInput = await promptUser(request.requestedSchema);
       *   return { action: 'accept', content: userInput };
       * });
       * ```
       */
      onRequest: async (serverName, handler) => {
        try {
          const internalClient = await this.getConnectedClientForServer(serverName);
          return internalClient.elicitation.onRequest(handler);
        } catch (err) {
          throw new error.MastraError(
            {
              id: "MCP_CLIENT_ON_REQUEST_ELICITATION_FAILED",
              domain: error.ErrorDomain.MCP,
              category: error.ErrorCategory.THIRD_PARTY,
              details: {
                serverName
              }
            },
            err
          );
        }
      }
    };
  }
  /**
   * Provides access to resource-related operations across all configured servers.
   *
   * Resources represent data exposed by MCP servers (files, database records, API responses, etc.).
   *
   * @example
   * ```typescript
   * // List all resources from all servers
   * const allResources = await mcp.resources.list();
   * Object.entries(allResources).forEach(([serverName, resources]) => {
   *   console.log(`${serverName}: ${resources.length} resources`);
   * });
   *
   * // Read a specific resource
   * const content = await mcp.resources.read('weatherServer', 'file://data.json');
   *
   * // Subscribe to resource updates
   * await mcp.resources.subscribe('weatherServer', 'file://data.json');
   * await mcp.resources.onUpdated('weatherServer', async (params) => {
   *   console.log(`Resource updated: ${params.uri}`);
   * });
   * ```
   */
  get resources() {
    this.addToInstanceCache();
    return {
      /**
       * Lists all available resources from all configured servers.
       *
       * Returns a map of server names to their resource arrays. Errors for individual
       * servers are logged but don't throw - failed servers return empty arrays.
       *
       * @returns Promise resolving to object mapping server names to resource arrays
       *
       * @example
       * ```typescript
       * const resources = await mcp.resources.list();
       * console.log(resources.weatherServer); // Array of resources
       * ```
       */
      list: async () => {
        const allResources = {};
        for (const serverName of Object.keys(this.serverConfigs)) {
          try {
            const internalClient = await this.getConnectedClientForServer(serverName);
            allResources[serverName] = await internalClient.resources.list();
          } catch (error$1) {
            const mastraError = new error.MastraError(
              {
                id: "MCP_CLIENT_LIST_RESOURCES_FAILED",
                domain: error.ErrorDomain.MCP,
                category: error.ErrorCategory.THIRD_PARTY,
                details: {
                  serverName
                }
              },
              error$1
            );
            this.logger.trackException(mastraError);
            this.logger.error("Failed to list resources from server:", { error: mastraError.toString() });
          }
        }
        return allResources;
      },
      /**
       * Lists all available resource templates from all configured servers.
       *
       * Resource templates are URI templates (RFC 6570) describing dynamic resources.
       * Errors for individual servers are logged but don't throw.
       *
       * @returns Promise resolving to object mapping server names to template arrays
       *
       * @example
       * ```typescript
       * const templates = await mcp.resources.templates();
       * console.log(templates.weatherServer); // Array of resource templates
       * ```
       */
      templates: async () => {
        const allTemplates = {};
        for (const serverName of Object.keys(this.serverConfigs)) {
          try {
            const internalClient = await this.getConnectedClientForServer(serverName);
            allTemplates[serverName] = await internalClient.resources.templates();
          } catch (error$1) {
            const mastraError = new error.MastraError(
              {
                id: "MCP_CLIENT_LIST_RESOURCE_TEMPLATES_FAILED",
                domain: error.ErrorDomain.MCP,
                category: error.ErrorCategory.THIRD_PARTY,
                details: {
                  serverName
                }
              },
              error$1
            );
            this.logger.trackException(mastraError);
            this.logger.error("Failed to list resource templates from server:", { error: mastraError.toString() });
          }
        }
        return allTemplates;
      },
      /**
       * Reads the content of a specific resource from a server.
       *
       * @param serverName - Name of the server to read from
       * @param uri - URI of the resource to read
       * @returns Promise resolving to the resource content
       * @throws {MastraError} If reading the resource fails
       *
       * @example
       * ```typescript
       * const content = await mcp.resources.read('weatherServer', 'file://config.json');
       * console.log(content.contents[0].text);
       * ```
       */
      read: async (serverName, uri) => {
        try {
          const internalClient = await this.getConnectedClientForServer(serverName);
          return internalClient.resources.read(uri);
        } catch (error$1) {
          throw new error.MastraError(
            {
              id: "MCP_CLIENT_READ_RESOURCE_FAILED",
              domain: error.ErrorDomain.MCP,
              category: error.ErrorCategory.THIRD_PARTY,
              details: {
                serverName,
                uri
              }
            },
            error$1
          );
        }
      },
      /**
       * Subscribes to updates for a specific resource on a server.
       *
       * @param serverName - Name of the server
       * @param uri - URI of the resource to subscribe to
       * @returns Promise resolving when subscription is established
       * @throws {MastraError} If subscription fails
       *
       * @example
       * ```typescript
       * await mcp.resources.subscribe('weatherServer', 'file://config.json');
       * ```
       */
      subscribe: async (serverName, uri) => {
        try {
          const internalClient = await this.getConnectedClientForServer(serverName);
          return internalClient.resources.subscribe(uri);
        } catch (error$1) {
          throw new error.MastraError(
            {
              id: "MCP_CLIENT_SUBSCRIBE_RESOURCE_FAILED",
              domain: error.ErrorDomain.MCP,
              category: error.ErrorCategory.THIRD_PARTY,
              details: {
                serverName,
                uri
              }
            },
            error$1
          );
        }
      },
      /**
       * Unsubscribes from updates for a specific resource on a server.
       *
       * @param serverName - Name of the server
       * @param uri - URI of the resource to unsubscribe from
       * @returns Promise resolving when unsubscription is complete
       * @throws {MastraError} If unsubscription fails
       *
       * @example
       * ```typescript
       * await mcp.resources.unsubscribe('weatherServer', 'file://config.json');
       * ```
       */
      unsubscribe: async (serverName, uri) => {
        try {
          const internalClient = await this.getConnectedClientForServer(serverName);
          return internalClient.resources.unsubscribe(uri);
        } catch (err) {
          throw new error.MastraError(
            {
              id: "MCP_CLIENT_UNSUBSCRIBE_RESOURCE_FAILED",
              domain: error.ErrorDomain.MCP,
              category: error.ErrorCategory.THIRD_PARTY,
              details: {
                serverName,
                uri
              }
            },
            err
          );
        }
      },
      /**
       * Sets a notification handler for when subscribed resources are updated on a server.
       *
       * @param serverName - Name of the server to monitor
       * @param handler - Callback function receiving the updated resource URI
       * @returns Promise resolving when handler is registered
       * @throws {MastraError} If setting up the handler fails
       *
       * @example
       * ```typescript
       * await mcp.resources.onUpdated('weatherServer', async (params) => {
       *   console.log(`Resource updated: ${params.uri}`);
       *   const content = await mcp.resources.read('weatherServer', params.uri);
       * });
       * ```
       */
      onUpdated: async (serverName, handler) => {
        try {
          const internalClient = await this.getConnectedClientForServer(serverName);
          return internalClient.resources.onUpdated(handler);
        } catch (err) {
          throw new error.MastraError(
            {
              id: "MCP_CLIENT_ON_UPDATED_RESOURCE_FAILED",
              domain: error.ErrorDomain.MCP,
              category: error.ErrorCategory.THIRD_PARTY,
              details: {
                serverName
              }
            },
            err
          );
        }
      },
      /**
       * Sets a notification handler for when the resource list changes on a server.
       *
       * @param serverName - Name of the server to monitor
       * @param handler - Callback function invoked when resources are added/removed
       * @returns Promise resolving when handler is registered
       * @throws {MastraError} If setting up the handler fails
       *
       * @example
       * ```typescript
       * await mcp.resources.onListChanged('weatherServer', async () => {
       *   console.log('Resource list changed, re-fetching...');
       *   const resources = await mcp.resources.list();
       * });
       * ```
       */
      onListChanged: async (serverName, handler) => {
        try {
          const internalClient = await this.getConnectedClientForServer(serverName);
          return internalClient.resources.onListChanged(handler);
        } catch (err) {
          throw new error.MastraError(
            {
              id: "MCP_CLIENT_ON_LIST_CHANGED_RESOURCE_FAILED",
              domain: error.ErrorDomain.MCP,
              category: error.ErrorCategory.THIRD_PARTY,
              details: {
                serverName
              }
            },
            err
          );
        }
      }
    };
  }
  /**
   * Provides access to prompt-related operations across all configured servers.
   *
   * Prompts are reusable message templates exposed by MCP servers that can be parameterized
   * and used for AI interactions.
   *
   * @example
   * ```typescript
   * // List all prompts from all servers
   * const allPrompts = await mcp.prompts.list();
   * Object.entries(allPrompts).forEach(([serverName, prompts]) => {
   *   console.log(`${serverName}: ${prompts.map(p => p.name).join(', ')}`);
   * });
   *
   * // Get a specific prompt with arguments
   * const prompt = await mcp.prompts.get({
   *   serverName: 'weatherServer',
   *   name: 'forecast-template',
   *   args: { city: 'London', days: 7 }
   * });
   * ```
   */
  get prompts() {
    this.addToInstanceCache();
    return {
      /**
       * Lists all available prompts from all configured servers.
       *
       * Returns a map of server names to their prompt arrays. Errors for individual
       * servers are logged but don't throw - failed servers return empty arrays.
       *
       * @returns Promise resolving to object mapping server names to prompt arrays
       *
       * @example
       * ```typescript
       * const prompts = await mcp.prompts.list();
       * console.log(prompts.weatherServer); // Array of prompts
       * ```
       */
      list: async () => {
        const allPrompts = {};
        for (const serverName of Object.keys(this.serverConfigs)) {
          try {
            const internalClient = await this.getConnectedClientForServer(serverName);
            allPrompts[serverName] = await internalClient.prompts.list();
          } catch (error$1) {
            const mastraError = new error.MastraError(
              {
                id: "MCP_CLIENT_LIST_PROMPTS_FAILED",
                domain: error.ErrorDomain.MCP,
                category: error.ErrorCategory.THIRD_PARTY,
                details: {
                  serverName
                }
              },
              error$1
            );
            this.logger.trackException(mastraError);
            this.logger.error("Failed to list prompts from server:", { error: mastraError.toString() });
          }
        }
        return allPrompts;
      },
      /**
       * Retrieves a specific prompt with its messages from a server.
       *
       * @param params - Parameters for the prompt request
       * @param params.serverName - Name of the server to retrieve from
       * @param params.name - Name of the prompt to retrieve
       * @param params.args - Optional arguments to populate the prompt template
       * @param params.version - Optional specific version of the prompt
       * @returns Promise resolving to the prompt result with messages
       * @throws {MastraError} If fetching the prompt fails
       *
       * @example
       * ```typescript
       * const prompt = await mcp.prompts.get({
       *   serverName: 'weatherServer',
       *   name: 'forecast',
       *   args: { city: 'London' },
       *   version: '1.0'
       * });
       * console.log(prompt.messages);
       * ```
       */
      get: async ({
        serverName,
        name,
        args,
        version
      }) => {
        try {
          const internalClient = await this.getConnectedClientForServer(serverName);
          return internalClient.prompts.get({ name, args, version });
        } catch (error$1) {
          throw new error.MastraError(
            {
              id: "MCP_CLIENT_GET_PROMPT_FAILED",
              domain: error.ErrorDomain.MCP,
              category: error.ErrorCategory.THIRD_PARTY,
              details: {
                serverName,
                name
              }
            },
            error$1
          );
        }
      },
      /**
       * Sets a notification handler for when the prompt list changes on a server.
       *
       * @param serverName - Name of the server to monitor
       * @param handler - Callback function invoked when prompts are added/removed/modified
       * @returns Promise resolving when handler is registered
       * @throws {MastraError} If setting up the handler fails
       *
       * @example
       * ```typescript
       * await mcp.prompts.onListChanged('weatherServer', async () => {
       *   console.log('Prompt list changed, re-fetching...');
       *   const prompts = await mcp.prompts.list();
       * });
       * ```
       */
      onListChanged: async (serverName, handler) => {
        try {
          const internalClient = await this.getConnectedClientForServer(serverName);
          return internalClient.prompts.onListChanged(handler);
        } catch (error$1) {
          throw new error.MastraError(
            {
              id: "MCP_CLIENT_ON_LIST_CHANGED_PROMPT_FAILED",
              domain: error.ErrorDomain.MCP,
              category: error.ErrorCategory.THIRD_PARTY,
              details: {
                serverName
              }
            },
            error$1
          );
        }
      }
    };
  }
  addToInstanceCache() {
    if (!mcpClientInstances.has(this.id)) {
      mcpClientInstances.set(this.id, this);
    }
  }
  makeId() {
    const text = JSON.stringify(this.serverConfigs).normalize("NFKC");
    const idNamespace = uuid.v5(`MCPClient`, uuid.v5.DNS);
    return uuid.v5(text, idNamespace);
  }
  /**
   * Disconnects from all MCP servers and cleans up resources.
   *
   * This method gracefully closes all server connections and clears internal caches.
   * Safe to call multiple times - subsequent calls will wait for the first disconnect to complete.
   *
   * @example
   * ```typescript
   * // Cleanup on application shutdown
   * process.on('SIGTERM', async () => {
   *   await mcp.disconnect();
   *   process.exit(0);
   * });
   * ```
   */
  async disconnect() {
    if (this.disconnectPromise) {
      return this.disconnectPromise;
    }
    this.disconnectPromise = (async () => {
      try {
        mcpClientInstances.delete(this.id);
        await Promise.allSettled(Array.from(this.mcpClientsById.values()).map((client) => client.disconnect()));
        this.mcpClientsById.clear();
      } finally {
        this.disconnectPromise = null;
      }
    })();
    return this.disconnectPromise;
  }
  /**
   * Retrieves all tools from all configured servers with namespaced names.
   *
   * Tool names are namespaced as `serverName_toolName` to prevent conflicts between servers.
   * This method is intended to be passed directly to an Agent definition.
   *
   * @returns Object mapping namespaced tool names to tool implementations.
   * Errors for individual servers are logged but don't throw - failed servers are skipped.
   *
   * @example
   * ```typescript
   * const agent = new Agent({
   *   id: 'multi-tool-agent',
   *   name: 'Multi-tool Agent',
   *   instructions: 'You have access to weather and stock tools.',
   *   model: 'openai/gpt-4',
   *   tools: await mcp.listTools(), // weather_getWeather, stockPrice_getPrice
   * });
   * ```
   */
  async listTools() {
    this.addToInstanceCache();
    const connectedTools = {};
    for (const serverName of Object.keys(this.serverConfigs)) {
      try {
        const client = await this.getConnectedClientForServer(serverName);
        const tools = await client.tools();
        for (const [toolName, toolConfig] of Object.entries(tools)) {
          connectedTools[`${serverName}_${toolName}`] = toolConfig;
        }
      } catch (error$1) {
        const mastraError = new error.MastraError(
          {
            id: "MCP_CLIENT_GET_TOOLS_FAILED",
            domain: error.ErrorDomain.MCP,
            category: error.ErrorCategory.THIRD_PARTY,
            details: {
              serverName
            }
          },
          error$1
        );
        this.logger.trackException(mastraError);
        this.logger.error("Failed to list tools from server:", { error: mastraError.toString() });
      }
    }
    return connectedTools;
  }
  /**
   * Returns toolsets organized by server name for dynamic tool injection.
   *
   * Unlike listTools(), this returns tools grouped by server without namespacing.
   * This is intended to be passed dynamically to the generate() or stream() method.
   *
   * @returns Object mapping server names to their tool collections.
   * Errors for individual servers are logged but don't throw - failed servers are skipped.
   *
   * @example
   * ```typescript
   * const agent = new Agent({
   *   id: 'dynamic-agent',
   *   name: 'Dynamic Agent',
   *   instructions: 'You can use tools dynamically.',
   *   model: 'openai/gpt-4',
   * });
   *
   * const response = await agent.stream(prompt, {
   *   toolsets: await mcp.listToolsets(), // { weather: {...}, stockPrice: {...} }
   * });
   * ```
   */
  async listToolsets() {
    this.addToInstanceCache();
    const connectedToolsets = {};
    for (const serverName of Object.keys(this.serverConfigs)) {
      try {
        const client = await this.getConnectedClientForServer(serverName);
        const tools = await client.tools();
        if (tools) {
          connectedToolsets[serverName] = tools;
        }
      } catch (error$1) {
        const mastraError = new error.MastraError(
          {
            id: "MCP_CLIENT_GET_TOOLSETS_FAILED",
            domain: error.ErrorDomain.MCP,
            category: error.ErrorCategory.THIRD_PARTY,
            details: {
              serverName
            }
          },
          error$1
        );
        this.logger.trackException(mastraError);
        this.logger.error("Failed to list toolsets from server:", { error: mastraError.toString() });
      }
    }
    return connectedToolsets;
  }
  /**
   * Gets current session IDs for all connected MCP clients using Streamable HTTP transport.
   *
   * Returns an object mapping server names to their session IDs. Only includes servers
   * that are currently connected via Streamable HTTP transport.
   *
   * @returns Object mapping server names to session IDs
   *
   * @example
   * ```typescript
   * const sessions = mcp.sessionIds;
   * console.log(sessions);
   * // { weatherServer: 'abc-123', stockServer: 'def-456' }
   * ```
   */
  get sessionIds() {
    const sessionIds = {};
    for (const [serverName, client] of this.mcpClientsById.entries()) {
      if (client.sessionId) {
        sessionIds[serverName] = client.sessionId;
      }
    }
    return sessionIds;
  }
  async getConnectedClient(name, config) {
    if (this.disconnectPromise) {
      await this.disconnectPromise;
    }
    const exists = this.mcpClientsById.has(name);
    const existingClient = this.mcpClientsById.get(name);
    this.logger.debug(`getConnectedClient ${name} exists: ${exists}`);
    if (exists) {
      if (!existingClient) {
        throw new Error(`Client ${name} exists but is undefined`);
      }
      await existingClient.connect();
      return existingClient;
    }
    this.logger.debug(`Connecting to ${name} MCP server`);
    const mcpClient = new InternalMastraMCPClient({
      name,
      server: config,
      timeout: config.timeout ?? this.defaultTimeout
    });
    mcpClient.__setLogger(this.logger);
    this.mcpClientsById.set(name, mcpClient);
    try {
      await mcpClient.connect();
    } catch (e) {
      const mastraError = new error.MastraError(
        {
          id: "MCP_CLIENT_CONNECT_FAILED",
          domain: error.ErrorDomain.MCP,
          category: error.ErrorCategory.THIRD_PARTY,
          text: `Failed to connect to MCP server ${name}: ${e instanceof Error ? e.stack || e.message : String(e)}`,
          details: {
            name
          }
        },
        e
      );
      this.logger.trackException(mastraError);
      this.logger.error("MCPClient errored connecting to MCP server:", { error: mastraError.toString() });
      this.mcpClientsById.delete(name);
      throw mastraError;
    }
    this.logger.debug(`Connected to ${name} MCP server`);
    return mcpClient;
  }
  async getConnectedClientForServer(serverName) {
    const serverConfig = this.serverConfigs[serverName];
    if (!serverConfig) {
      throw new Error(`Server configuration not found for name: ${serverName}`);
    }
    return this.getConnectedClient(serverName, serverConfig);
  }
};

// src/client/oauth-provider.ts
var InMemoryOAuthStorage = class {
  data = /* @__PURE__ */ new Map();
  set(key, value) {
    this.data.set(key, value);
  }
  get(key) {
    return this.data.get(key);
  }
  delete(key) {
    this.data.delete(key);
  }
  clear() {
    this.data.clear();
  }
};
var MCPOAuthClientProvider = class {
  _redirectUrl;
  _clientMetadata;
  storage;
  onRedirect;
  generateState;
  _clientInfo;
  constructor(options) {
    this._redirectUrl = options.redirectUrl;
    this._clientMetadata = options.clientMetadata;
    this._clientInfo = options.clientInformation;
    this.storage = options.storage ?? new InMemoryOAuthStorage();
    this.onRedirect = options.onRedirectToAuthorization;
    this.generateState = options.stateGenerator ?? (() => crypto.randomUUID());
  }
  /**
   * The URL to redirect the user agent to after authorization.
   */
  get redirectUrl() {
    return this._redirectUrl;
  }
  /**
   * Metadata about this OAuth client.
   */
  get clientMetadata() {
    return this._clientMetadata;
  }
  /**
   * Returns a OAuth2 state parameter.
   */
  async state() {
    return this.generateState();
  }
  /**
   * Loads information about this OAuth client.
   */
  async clientInformation() {
    if (this._clientInfo) {
      return this._clientInfo;
    }
    const stored = await this.storage.get("client_info");
    if (stored) {
      try {
        return JSON.parse(stored);
      } catch {
      }
    }
    return void 0;
  }
  /**
   * Saves dynamically registered client information.
   */
  async saveClientInformation(clientInformation) {
    this._clientInfo = clientInformation;
    await this.storage.set("client_info", JSON.stringify(clientInformation));
  }
  /**
   * Loads existing OAuth tokens.
   */
  async tokens() {
    const stored = await this.storage.get("tokens");
    if (stored) {
      try {
        return JSON.parse(stored);
      } catch {
      }
    }
    return void 0;
  }
  /**
   * Stores new OAuth tokens after successful authorization.
   */
  async saveTokens(tokens) {
    await this.storage.set("tokens", JSON.stringify(tokens));
  }
  /**
   * Invoked to redirect the user agent to the authorization URL.
   */
  async redirectToAuthorization(authorizationUrl) {
    if (this.onRedirect) {
      await this.onRedirect(authorizationUrl);
    } else {
      console.info(`Authorization required. Please visit: ${authorizationUrl.toString()}`);
    }
  }
  /**
   * Saves a PKCE code verifier before redirecting to authorization.
   */
  async saveCodeVerifier(codeVerifier) {
    await this.storage.set("code_verifier", codeVerifier);
  }
  /**
   * Loads the PKCE code verifier for validating authorization result.
   */
  async codeVerifier() {
    const verifier = await this.storage.get("code_verifier");
    if (!verifier) {
      throw new Error("No code verifier found. Authorization flow may not have started properly.");
    }
    return verifier;
  }
  /**
   * Optional: Custom client authentication for token requests.
   * Uses default behavior if not implemented.
   */
  async addClientAuthentication(_headers, _params, _url, _metadata) {
  }
  /**
   * Invalidate credentials when server indicates they're no longer valid.
   */
  async invalidateCredentials(scope) {
    switch (scope) {
      case "all":
        await this.storage.delete("tokens");
        await this.storage.delete("client_info");
        await this.storage.delete("code_verifier");
        this._clientInfo = void 0;
        break;
      case "client":
        await this.storage.delete("client_info");
        this._clientInfo = void 0;
        break;
      case "tokens":
        await this.storage.delete("tokens");
        break;
      case "verifier":
        await this.storage.delete("code_verifier");
        break;
    }
  }
  /**
   * Clear all stored OAuth data.
   * Useful for logging out or resetting state.
   */
  async clear() {
    await this.invalidateCredentials("all");
  }
  /**
   * Check if the provider has valid (non-expired) tokens.
   */
  async hasValidTokens() {
    const currentTokens = await this.tokens();
    if (!currentTokens) return false;
    if (!currentTokens.access_token) return false;
    return true;
  }
};
function createSimpleTokenProvider(accessToken, options) {
  const tokens = {
    access_token: accessToken,
    token_type: options.tokenType ?? "Bearer",
    refresh_token: options.refreshToken,
    expires_in: options.expiresIn,
    scope: options.scope
  };
  const storage = new InMemoryOAuthStorage();
  storage.set("tokens", JSON.stringify(tokens));
  if (options.clientInformation) {
    storage.set("client_info", JSON.stringify(options.clientInformation));
  }
  return new MCPOAuthClientProvider({
    redirectUrl: options.redirectUrl,
    clientMetadata: options.clientMetadata,
    clientInformation: options.clientInformation,
    storage
  });
}

// ../../node_modules/.pnpm/hono@4.11.9/node_modules/hono/dist/utils/stream.js
var StreamingApi = class {
  writer;
  encoder;
  writable;
  abortSubscribers = [];
  responseReadable;
  /**
   * Whether the stream has been aborted.
   */
  aborted = false;
  /**
   * Whether the stream has been closed normally.
   */
  closed = false;
  constructor(writable, _readable) {
    this.writable = writable;
    this.writer = writable.getWriter();
    this.encoder = new TextEncoder();
    const reader = _readable.getReader();
    this.abortSubscribers.push(async () => {
      await reader.cancel();
    });
    this.responseReadable = new ReadableStream({
      async pull(controller) {
        const { done, value } = await reader.read();
        done ? controller.close() : controller.enqueue(value);
      },
      cancel: () => {
        this.abort();
      }
    });
  }
  async write(input) {
    try {
      if (typeof input === "string") {
        input = this.encoder.encode(input);
      }
      await this.writer.write(input);
    } catch {
    }
    return this;
  }
  async writeln(input) {
    await this.write(input + "\n");
    return this;
  }
  sleep(ms) {
    return new Promise((res) => setTimeout(res, ms));
  }
  async close() {
    try {
      await this.writer.close();
    } catch {
    }
    this.closed = true;
  }
  async pipe(body) {
    this.writer.releaseLock();
    await body.pipeTo(this.writable, { preventClose: true });
    this.writer = this.writable.getWriter();
  }
  onAbort(listener) {
    this.abortSubscribers.push(listener);
  }
  /**
   * Abort the stream.
   * You can call this method when stream is aborted by external event.
   */
  abort() {
    if (!this.aborted) {
      this.aborted = true;
      this.abortSubscribers.forEach((subscriber) => subscriber());
    }
  }
};

// ../../node_modules/.pnpm/hono@4.11.9/node_modules/hono/dist/helper/streaming/utils.js
var isOldBunVersion = () => {
  const version = typeof Bun !== "undefined" ? Bun.version : void 0;
  if (version === void 0) {
    return false;
  }
  const result = version.startsWith("1.1") || version.startsWith("1.0") || version.startsWith("0.");
  isOldBunVersion = () => result;
  return result;
};

// ../../node_modules/.pnpm/hono@4.11.9/node_modules/hono/dist/utils/html.js
var HtmlEscapedCallbackPhase = {
  Stringify: 1};
var resolveCallback = async (str, phase, preserveCallbacks, context, buffer) => {
  if (typeof str === "object" && !(str instanceof String)) {
    if (!(str instanceof Promise)) {
      str = str.toString();
    }
    if (str instanceof Promise) {
      str = await str;
    }
  }
  const callbacks = str.callbacks;
  if (!callbacks?.length) {
    return Promise.resolve(str);
  }
  if (buffer) {
    buffer[0] += str;
  } else {
    buffer = [str];
  }
  const resStr = Promise.all(callbacks.map((c) => c({ phase, buffer, context }))).then(
    (res) => Promise.all(
      res.filter(Boolean).map((str2) => resolveCallback(str2, phase, false, context, buffer))
    ).then(() => buffer[0])
  );
  {
    return resStr;
  }
};

// ../../node_modules/.pnpm/hono@4.11.9/node_modules/hono/dist/helper/streaming/sse.js
var SSEStreamingApi = class extends StreamingApi {
  constructor(writable, readable) {
    super(writable, readable);
  }
  async writeSSE(message) {
    const data = await resolveCallback(message.data, HtmlEscapedCallbackPhase.Stringify, false, {});
    const dataLines = data.split(/\r\n|\r|\n/).map((line) => {
      return `data: ${line}`;
    }).join("\n");
    const sseData = [
      message.event && `event: ${message.event}`,
      dataLines,
      message.id && `id: ${message.id}`,
      message.retry && `retry: ${message.retry}`
    ].filter(Boolean).join("\n") + "\n\n";
    await this.write(sseData);
  }
};
var run = async (stream2, cb, onError) => {
  try {
    await cb(stream2);
  } catch (e) {
    {
      console.error(e);
    }
  } finally {
    stream2.close();
  }
};
var contextStash = /* @__PURE__ */ new WeakMap();
var streamSSE = (c, cb, onError) => {
  const { readable, writable } = new TransformStream();
  const stream2 = new SSEStreamingApi(writable, readable);
  if (isOldBunVersion()) {
    c.req.raw.signal.addEventListener("abort", () => {
      if (!stream2.closed) {
        stream2.abort();
      }
    });
  }
  contextStash.set(stream2.responseReadable, c);
  c.header("Transfer-Encoding", "chunked");
  c.header("Content-Type", "text/event-stream");
  c.header("Cache-Control", "no-cache");
  c.header("Connection", "keep-alive");
  run(stream2, cb);
  return c.newResponse(stream2.responseReadable);
};
var MAXIMUM_MESSAGE_SIZE = 4 * 1024 * 1024;
var SSETransport = class {
  messageUrl;
  stream;
  _sessionId;
  onclose;
  onerror;
  onmessage;
  /**
   * Creates a new SSETransport, which will direct the MPC client to POST messages to messageUrl
   */
  constructor(messageUrl, stream2) {
    this.messageUrl = messageUrl;
    this.stream = stream2;
    this._sessionId = crypto.randomUUID();
    this.stream.onAbort(() => {
      void this.close();
    });
  }
  get sessionId() {
    return this._sessionId;
  }
  // start() is automatically called after MCP Server connects to the transport
  async start() {
    if (this.stream == null) {
      throw new Error("Stream not initialized");
    }
    if (this.stream.closed) {
      throw new Error("SSE transport already closed!");
    }
    await this.stream.writeSSE({
      event: "ping",
      data: ""
    });
    await this.stream.writeSSE({
      event: "endpoint",
      data: `${this.messageUrl}?sessionId=${this.sessionId}`
    });
  }
  async handlePostMessage(context) {
    if (this.stream?.closed == null) {
      return context.text("SSE connection not established", 500);
    }
    try {
      const contentType = context.req.header("content-type") || "";
      if (!contentType.includes("application/json")) {
        throw new Error(`Unsupported content-type: ${contentType}`);
      }
      const contentLength = Number.parseInt(context.req.header("content-length") || "0", 10);
      if (contentLength > MAXIMUM_MESSAGE_SIZE) {
        throw new Error(`Request body too large: ${contentLength} bytes`);
      }
      const body = await context.req.json();
      await this.handleMessage(body);
      return context.text("Accepted", 202);
    } catch (error) {
      this.onerror?.(error);
      return context.text("Error", 400);
    }
  }
  /**
   * Handle a client message, regardless of how it arrived. This can be used to inform the server of messages that arrive via a means different than HTTP POST.
   */
  async handleMessage(message) {
    let parsedMessage;
    try {
      parsedMessage = types_js.JSONRPCMessageSchema.parse(message);
    } catch (error) {
      this.onerror?.(error);
      throw error;
    }
    this.onmessage?.(parsedMessage);
  }
  async close() {
    if (this.stream?.closed) {
      this.stream.abort();
    }
  }
  async send(message) {
    if (this.stream?.closed) {
      throw new Error("Not connected");
    }
    await this.stream.writeSSE({
      event: "message",
      data: JSON.stringify(message)
    });
  }
};
var ServerPromptActions = class {
  getLogger;
  getSdkServer;
  clearDefinedPrompts;
  /**
   * @internal
   */
  constructor(dependencies) {
    this.getLogger = dependencies.getLogger;
    this.getSdkServer = dependencies.getSdkServer;
    this.clearDefinedPrompts = dependencies.clearDefinedPrompts;
  }
  /**
   * Notifies clients that the overall list of available prompts has changed.
   *
   * This clears the internal prompt cache and sends a `notifications/prompts/list_changed`
   * message to all clients, prompting them to re-fetch the prompt list.
   *
   * @throws {MastraError} If sending the notification fails
   *
   * @example
   * ```typescript
   * // After adding or modifying prompts
   * await server.prompts.notifyListChanged();
   * ```
   */
  async notifyListChanged() {
    this.getLogger().info("Prompt list change externally notified. Clearing definedPrompts and sending notification.");
    this.clearDefinedPrompts();
    try {
      await this.getSdkServer().sendPromptListChanged();
    } catch (error$1) {
      const mastraError = new error.MastraError(
        {
          id: "MCP_SERVER_PROMPT_LIST_CHANGED_NOTIFICATION_FAILED",
          domain: error.ErrorDomain.MCP,
          category: error.ErrorCategory.THIRD_PARTY,
          text: "Failed to send prompt list changed notification"
        },
        error$1
      );
      this.getLogger().error("Failed to send prompt list changed notification:", {
        error: mastraError.toString()
      });
      this.getLogger().trackException(mastraError);
      throw mastraError;
    }
  }
};
var ServerResourceActions = class {
  getSubscriptions;
  getLogger;
  getSdkServer;
  clearDefinedResources;
  clearDefinedResourceTemplates;
  /**
   * @internal
   */
  constructor(dependencies) {
    this.getSubscriptions = dependencies.getSubscriptions;
    this.getLogger = dependencies.getLogger;
    this.getSdkServer = dependencies.getSdkServer;
    this.clearDefinedResources = dependencies.clearDefinedResources;
    this.clearDefinedResourceTemplates = dependencies.clearDefinedResourceTemplates;
  }
  /**
   * Notifies subscribed clients that a specific resource has been updated.
   *
   * If clients are subscribed to the resource URI, they will receive a
   * `notifications/resources/updated` message to re-fetch the resource content.
   *
   * @param params - Notification parameters
   * @param params.uri - URI of the resource that was updated
   * @throws {MastraError} If sending the notification fails
   *
   * @example
   * ```typescript
   * // After updating a file resource
   * await server.resources.notifyUpdated({ uri: 'file://data.txt' });
   * ```
   */
  async notifyUpdated({ uri }) {
    if (this.getSubscriptions().has(uri)) {
      this.getLogger().info(`Sending notifications/resources/updated for externally notified resource: ${uri}`);
      try {
        await this.getSdkServer().sendResourceUpdated({ uri });
      } catch (error$1) {
        const mastraError = new error.MastraError(
          {
            id: "MCP_SERVER_RESOURCE_UPDATED_NOTIFICATION_FAILED",
            domain: error.ErrorDomain.MCP,
            category: error.ErrorCategory.THIRD_PARTY,
            text: "Failed to send resource updated notification",
            details: {
              uri
            }
          },
          error$1
        );
        this.getLogger().trackException(mastraError);
        this.getLogger().error("Failed to send resource updated notification:", {
          error: mastraError.toString()
        });
        throw mastraError;
      }
    } else {
      this.getLogger().debug(`Resource ${uri} was updated, but no active subscriptions for it.`);
    }
  }
  /**
   * Notifies clients that the overall list of available resources has changed.
   *
   * This clears the internal resource cache and sends a `notifications/resources/list_changed`
   * message to all clients, prompting them to re-fetch the resource list.
   *
   * @throws {MastraError} If sending the notification fails
   *
   * @example
   * ```typescript
   * // After adding a new resource to your resource handler
   * await server.resources.notifyListChanged();
   * ```
   */
  async notifyListChanged() {
    this.getLogger().info(
      "Resource list change externally notified. Clearing definedResources and sending notification."
    );
    this.clearDefinedResources();
    this.clearDefinedResourceTemplates();
    try {
      await this.getSdkServer().sendResourceListChanged();
    } catch (error$1) {
      const mastraError = new error.MastraError(
        {
          id: "MCP_SERVER_RESOURCE_LIST_CHANGED_NOTIFICATION_FAILED",
          domain: error.ErrorDomain.MCP,
          category: error.ErrorCategory.THIRD_PARTY,
          text: "Failed to send resource list changed notification"
        },
        error$1
      );
      this.getLogger().trackException(mastraError);
      this.getLogger().error("Failed to send resource list changed notification:", {
        error: mastraError.toString()
      });
      throw mastraError;
    }
  }
};

// src/server/server.ts
var MCPServer = class extends mcp.MCPServerBase {
  server;
  stdioTransport;
  sseTransport;
  sseHonoTransports;
  streamableHTTPTransports = /* @__PURE__ */ new Map();
  // Track server instances for each HTTP session
  httpServerInstances = /* @__PURE__ */ new Map();
  definedResources;
  definedResourceTemplates;
  resourceOptions;
  definedPrompts;
  promptOptions;
  subscriptions = /* @__PURE__ */ new Set();
  currentLoggingLevel;
  /**
   * Provides methods to notify clients about resource changes.
   *
   * @example
   * ```typescript
   * // Notify that a specific resource was updated
   * await server.resources.notifyUpdated({ uri: 'file://data.txt' });
   *
   * // Notify that the resource list changed
   * await server.resources.notifyListChanged();
   * ```
   */
  resources;
  /**
   * Provides methods to notify clients about prompt changes.
   *
   * @example
   * ```typescript
   * // Notify that the prompt list changed
   * await server.prompts.notifyListChanged();
   * ```
   */
  prompts;
  /**
   * Provides methods for interactive user input collection during tool execution.
   *
   * @example
   * ```typescript
   * // Within a tool's execute function
   * const result = await options.elicitation.sendRequest({
   *   message: 'Please provide your email address',
   *   requestedSchema: {
   *     type: 'object',
   *     properties: {
   *       email: { type: 'string', format: 'email' }
   *     },
   *     required: ['email']
   *   }
   * });
   * ```
   */
  elicitation;
  /**
   * Gets the stdio transport instance if the server was started using stdio.
   *
   * This is primarily for internal checks or testing purposes.
   *
   * @returns The stdio transport instance, or undefined if not using stdio transport
   */
  getStdioTransport() {
    return this.stdioTransport;
  }
  /**
   * Gets the SSE transport instance if the server was started using SSE.
   *
   * This is primarily for internal checks or testing purposes.
   *
   * @returns The SSE transport instance, or undefined if not using SSE transport
   */
  getSseTransport() {
    return this.sseTransport;
  }
  /**
   * Gets the Hono SSE transport instance for a specific session.
   *
   * This is primarily for internal checks or testing purposes.
   *
   * @param sessionId - The session identifier
   * @returns The Hono SSE transport instance, or undefined if session not found
   */
  getSseHonoTransport(sessionId) {
    return this.sseHonoTransports.get(sessionId);
  }
  /**
   * Gets the underlying MCP SDK Server instance.
   *
   * This provides access to the low-level server instance for advanced use cases.
   *
   * @returns The Server instance from @modelcontextprotocol/sdk
   */
  getServer() {
    return this.server;
  }
  /**
   * Creates a new MCPServer instance.
   *
   * The server exposes tools, agents, and workflows to MCP clients. Agents are automatically
   * converted to tools named `ask_<agentKey>`, and workflows become tools named `run_<workflowKey>`.
   *
   * @param opts - Configuration options for the server
   * @param opts.name - Descriptive name for the server (e.g., 'My Weather Server')
   * @param opts.version - Semantic version of the server (e.g., '1.0.0')
   * @param opts.tools - Object mapping tool names to tool definitions
   * @param opts.agents - Optional object mapping agent identifiers to Agent instances
   * @param opts.workflows - Optional object mapping workflow identifiers to Workflow instances
   * @param opts.resources - Optional resource configuration for exposing data and content
   * @param opts.prompts - Optional prompt configuration for exposing reusable templates
   * @param opts.id - Optional unique identifier (generated if not provided)
   * @param opts.description - Optional description of what the server does
   *
   * @example
   * ```typescript
   * import { MCPServer } from '@mastra/mcp';
   * import { Agent } from '@mastra/core/agent';
   * import { createTool } from '@mastra/core/tools';
   * import { z } from 'zod';
   *
   * const myAgent = new Agent({
   *   id: 'helper',
   *   name: 'Helper Agent',
   *   description: 'A helpful assistant',
   *   instructions: 'You are helpful.',
   *   model: 'openai/gpt-4o-mini',
   * });
   *
   * const server = new MCPServer({
   *   name: 'My Server',
   *   version: '1.0.0',
   *   tools: {
   *     weatherTool: createTool({
   *       id: 'getWeather',
   *       description: 'Gets weather',
   *       inputSchema: z.object({ location: z.string() }),
   *       execute: async (inputData) => `Sunny in ${inputData.location}`,
   *     })
   *   },
   *   agents: { myAgent },
   * });
   * ```
   */
  constructor(opts) {
    super(opts);
    this.resourceOptions = opts.resources;
    this.promptOptions = opts.prompts;
    const capabilities = {
      tools: {},
      logging: { enabled: true },
      elicitation: {}
    };
    if (opts.resources) {
      capabilities.resources = { subscribe: true, listChanged: true };
    }
    if (opts.prompts) {
      capabilities.prompts = { listChanged: true };
    }
    this.server = new index_js$1.Server(
      {
        name: this.name,
        version: this.version
      },
      {
        capabilities,
        ...this.instructions ? { instructions: this.instructions } : {}
      }
    );
    this.logger.info(
      `Initialized MCPServer '${this.name}' v${this.version} (ID: ${this.id}) with tools: ${Object.keys(this.convertedTools).join(", ")} and resources. Capabilities: ${JSON.stringify(capabilities)}`
    );
    this.sseHonoTransports = /* @__PURE__ */ new Map();
    this.registerHandlersOnServer(this.server);
    this.resources = new ServerResourceActions({
      getSubscriptions: () => this.subscriptions,
      getLogger: () => this.logger,
      getSdkServer: () => this.server,
      clearDefinedResources: () => {
        this.definedResources = void 0;
      },
      clearDefinedResourceTemplates: () => {
        this.definedResourceTemplates = void 0;
      }
    });
    this.prompts = new ServerPromptActions({
      getLogger: () => this.logger,
      getSdkServer: () => this.server,
      clearDefinedPrompts: () => {
        this.definedPrompts = void 0;
      }
    });
    this.elicitation = {
      sendRequest: async (request, options) => {
        return this.handleElicitationRequest(request, void 0, options);
      }
    };
  }
  /**
   * Handle an elicitation request by sending it to the connected client.
   * This method sends an elicitation/create request to the client and waits for the response.
   *
   * @param request - The elicitation request containing message and schema
   * @param serverInstance - Optional server instance to use; defaults to main server for backward compatibility
   * @param options - Optional request options (timeout, signal, etc.)
   * @returns Promise that resolves to the client's response
   */
  async handleElicitationRequest(request, serverInstance, options) {
    this.logger.debug(`Sending elicitation request: ${request.message}`);
    const server = serverInstance || this.server;
    const response = await server.elicitInput(request, options);
    this.logger.debug(`Received elicitation response: ${JSON.stringify(response)}`);
    return response;
  }
  /**
   * Reads and parses the JSON body from an HTTP request.
   * If the request body was already parsed by middleware (e.g., express.json()),
   * it uses the pre-parsed body from req.body. Otherwise, it reads from the stream.
   *
   * This allows the MCP server to work with Express apps that use express.json()
   * globally without requiring special route exclusions.
   *
   * @param req - The incoming HTTP request
   * @param options - Optional configuration
   * @param options.preParsedOnly - If true, only return pre-parsed body from middleware,
   *   returning undefined if not available. This allows the caller to fall back to
   *   their own body reading logic (e.g., SDK's getRawBody with size limits).
   */
  async readJsonBody(req, options) {
    const reqWithBody = req;
    if (reqWithBody.body !== void 0) {
      return reqWithBody.body;
    }
    if (options?.preParsedOnly) {
      return void 0;
    }
    return new Promise((resolve, reject) => {
      let data = "";
      req.on("data", (chunk) => data += chunk);
      req.on("end", () => {
        try {
          resolve(JSON.parse(data));
        } catch (e) {
          reject(e);
        }
      });
      req.on("error", reject);
    });
  }
  /**
   * Creates a new Server instance configured with all handlers for HTTP sessions.
   * Each HTTP client connection gets its own Server instance to avoid routing conflicts.
   */
  createServerInstance() {
    const capabilities = {
      tools: {},
      logging: { enabled: true },
      elicitation: {}
    };
    if (this.resourceOptions) {
      capabilities.resources = { subscribe: true, listChanged: true };
    }
    if (this.promptOptions) {
      capabilities.prompts = { listChanged: true };
    }
    const serverInstance = new index_js$1.Server(
      {
        name: this.name,
        version: this.version
      },
      {
        capabilities,
        ...this.instructions ? { instructions: this.instructions } : {}
      }
    );
    this.registerHandlersOnServer(serverInstance);
    return serverInstance;
  }
  /**
   * Registers all MCP handlers on a given server instance.
   * This allows us to create multiple server instances with identical functionality.
   */
  registerHandlersOnServer(serverInstance) {
    serverInstance.setRequestHandler(types_js.ListToolsRequestSchema, async () => {
      this.logger.debug("Handling ListTools request");
      return {
        tools: Object.values(this.convertedTools).map((tool) => {
          const toolSpec = {
            name: tool.id || "unknown",
            description: tool.description,
            inputSchema: tool.parameters.jsonSchema
          };
          if (tool.outputSchema) {
            toolSpec.outputSchema = tool.outputSchema.jsonSchema;
          }
          if (tool.mcp?.annotations) {
            toolSpec.annotations = tool.mcp.annotations;
          }
          if (tool.mcp?._meta) {
            toolSpec._meta = tool.mcp._meta;
          }
          return toolSpec;
        })
      };
    });
    serverInstance.setRequestHandler(types_js.CallToolRequestSchema, async (request, extra) => {
      const startTime = Date.now();
      try {
        const tool = this.convertedTools[request.params.name];
        if (!tool) {
          this.logger.warn(`CallTool: Unknown tool '${request.params.name}' requested.`);
          return {
            content: [{ type: "text", text: `Unknown tool: ${request.params.name}` }],
            isError: true
          };
        }
        const validation = tool.parameters.validate?.(request.params.arguments ?? {});
        if (validation && !validation.success) {
          this.logger.warn(`CallTool: Invalid tool arguments for '${request.params.name}'`, {
            errors: validation.error
          });
          let errorMessages = "Validation failed";
          if ("errors" in validation.error && Array.isArray(validation.error.errors)) {
            errorMessages = validation.error.errors.map((e) => `- ${e.path?.join(".") || "root"}: ${e.message}`).join("\n");
          } else if (validation.error instanceof Error) {
            errorMessages = validation.error.message;
          }
          return {
            content: [
              {
                type: "text",
                text: `Tool validation failed. Please fix the following errors and try again:
${errorMessages}

Provided arguments: ${JSON.stringify(request.params.arguments, null, 2)}`
              }
            ],
            isError: true
            // Set to true so the LLM sees the error and can self-correct
          };
        }
        if (!tool.execute) {
          this.logger.warn(`CallTool: Tool '${request.params.name}' does not have an execute function.`);
          return {
            content: [{ type: "text", text: `Tool '${request.params.name}' does not have an execute function.` }],
            isError: true
          };
        }
        const sessionElicitation = {
          sendRequest: async (request2, options) => {
            return this.handleElicitationRequest(request2, serverInstance, options);
          }
        };
        const mcpOptions = {
          messages: [],
          toolCallId: "",
          // Pass MCP-specific context through the mcp property
          mcp: {
            elicitation: sessionElicitation,
            extra
          },
          // @ts-expect-error this is to let people know that the elicitation and extra keys are now nested under mcp.elicitation and mcp.extra in tool arguments
          get elicitation() {
            throw new Error(`The "elicitation" key is now nested under "mcp.elicitation" in tool arguments`);
          },
          get extra() {
            throw new Error(`The "extra" key is now nested under "mcp.extra" in tool arguments`);
          }
        };
        const result = await tool.execute(validation?.value ?? request.params.arguments ?? {}, mcpOptions);
        this.logger.debug(`CallTool: Tool '${request.params.name}' executed successfully with result:`, result);
        const duration = Date.now() - startTime;
        this.logger.info(`Tool '${request.params.name}' executed successfully in ${duration}ms.`);
        const response = { isError: false, content: [] };
        if (tool.outputSchema) {
          let structuredContent;
          if (result && typeof result === "object" && "structuredContent" in result) {
            structuredContent = result.structuredContent;
          } else {
            structuredContent = result;
          }
          const outputValidation = tool.outputSchema.validate?.(structuredContent ?? {});
          if (outputValidation && !outputValidation.success) {
            this.logger.warn(`CallTool: Invalid structured content for '${request.params.name}'`, {
              errors: outputValidation.error
            });
            throw new Error(
              `Invalid structured content for tool ${request.params.name}: ${JSON.stringify(outputValidation.error)}`
            );
          }
          response.structuredContent = structuredContent;
        }
        if (response.structuredContent) {
          response.content = [{ type: "text", text: JSON.stringify(response.structuredContent) }];
        } else {
          response.content = [
            {
              type: "text",
              text: typeof result === "string" ? result : JSON.stringify(result)
            }
          ];
        }
        return response;
      } catch (error$1) {
        const duration = Date.now() - startTime;
        if (error$1 instanceof zod.z.ZodError) {
          this.logger.warn("Invalid tool arguments", {
            tool: request.params.name,
            errors: error$1.errors,
            duration: `${duration}ms`
          });
          return {
            content: [
              {
                type: "text",
                text: `Invalid arguments: ${error$1.errors.map((e) => `${e.path.join(".")}: ${e.message}`).join(", ")}`
              }
            ],
            isError: true
          };
        }
        this.logger.error(`Tool execution failed: ${request.params.name}`, { error: error$1 });
        if (error$1 instanceof error.MastraError) {
          return {
            content: [{ type: "text", text: JSON.stringify(error$1.toJSON()) }],
            isError: true
          };
        }
        return {
          content: [{ type: "text", text: `Error: ${error$1 instanceof Error ? error$1.message : String(error$1)}` }],
          isError: true
        };
      }
    });
    serverInstance.setRequestHandler(types_js.SetLevelRequestSchema, async (request) => {
      this.currentLoggingLevel = request.params.level;
      this.logger.debug(`Logging level set to: ${request.params.level}`);
      return {};
    });
    if (this.resourceOptions) {
      this.registerResourceHandlersOnServer(serverInstance);
    }
    if (this.promptOptions) {
      this.registerPromptHandlersOnServer(serverInstance);
    }
  }
  /**
   * Registers resource-related handlers on a server instance.
   */
  registerResourceHandlersOnServer(serverInstance) {
    const capturedResourceOptions = this.resourceOptions;
    if (!capturedResourceOptions) return;
    if (capturedResourceOptions.listResources) {
      serverInstance.setRequestHandler(types_js.ListResourcesRequestSchema, async (_request, extra) => {
        this.logger.debug("Handling ListResources request");
        if (this.definedResources) {
          return { resources: this.definedResources };
        } else {
          try {
            const resources = await capturedResourceOptions.listResources({ extra });
            this.definedResources = resources;
            this.logger.debug(`Fetched and cached ${this.definedResources.length} resources.`);
            return { resources: this.definedResources };
          } catch (error) {
            this.logger.error("Error fetching resources via listResources():", { error });
            throw error;
          }
        }
      });
    }
    if (capturedResourceOptions.getResourceContent) {
      serverInstance.setRequestHandler(types_js.ReadResourceRequestSchema, async (request, extra) => {
        const startTime = Date.now();
        const uri = request.params.uri;
        this.logger.debug(`Handling ReadResource request for URI: ${uri}`);
        if (!this.definedResources) {
          const resources = await this.resourceOptions?.listResources?.({ extra });
          if (!resources) throw new Error("Failed to load resources");
          this.definedResources = resources;
        }
        const resource = this.definedResources?.find((r) => r.uri === uri);
        if (!resource) {
          this.logger.warn(`ReadResource: Unknown resource URI '${uri}' requested.`);
          throw new Error(`Resource not found: ${uri}`);
        }
        try {
          const resourcesOrResourceContent = await capturedResourceOptions.getResourceContent({ uri, extra });
          const resourcesContent = Array.isArray(resourcesOrResourceContent) ? resourcesOrResourceContent : [resourcesOrResourceContent];
          const contents = resourcesContent.map((resourceContent) => {
            const contentItem = {
              uri: resource.uri,
              mimeType: resource.mimeType
            };
            if ("text" in resourceContent) {
              contentItem.text = resourceContent.text;
            }
            if ("blob" in resourceContent) {
              contentItem.blob = resourceContent.blob;
            }
            return contentItem;
          });
          const duration = Date.now() - startTime;
          this.logger.info(`Resource '${uri}' read successfully in ${duration}ms.`);
          return {
            contents
          };
        } catch (error) {
          const duration = Date.now() - startTime;
          this.logger.error(`Failed to get content for resource URI '${uri}' in ${duration}ms`, { error });
          throw error;
        }
      });
    }
    if (capturedResourceOptions.resourceTemplates) {
      serverInstance.setRequestHandler(types_js.ListResourceTemplatesRequestSchema, async (_request, extra) => {
        this.logger.debug("Handling ListResourceTemplates request");
        if (this.definedResourceTemplates) {
          return { resourceTemplates: this.definedResourceTemplates };
        } else {
          try {
            const templates = await capturedResourceOptions.resourceTemplates({ extra });
            this.definedResourceTemplates = templates;
            this.logger.debug(`Fetched and cached ${this.definedResourceTemplates.length} resource templates.`);
            return { resourceTemplates: this.definedResourceTemplates };
          } catch (error) {
            this.logger.error("Error fetching resource templates via resourceTemplates():", { error });
            throw error;
          }
        }
      });
    }
    serverInstance.setRequestHandler(types_js.SubscribeRequestSchema, async (request) => {
      const uri = request.params.uri;
      this.logger.info(`Received resources/subscribe request for URI: ${uri}`);
      this.subscriptions.add(uri);
      return {};
    });
    serverInstance.setRequestHandler(types_js.UnsubscribeRequestSchema, async (request) => {
      const uri = request.params.uri;
      this.logger.info(`Received resources/unsubscribe request for URI: ${uri}`);
      this.subscriptions.delete(uri);
      return {};
    });
  }
  /**
   * Registers prompt-related handlers on a server instance.
   */
  registerPromptHandlersOnServer(serverInstance) {
    const capturedPromptOptions = this.promptOptions;
    if (!capturedPromptOptions) return;
    if (capturedPromptOptions.listPrompts) {
      serverInstance.setRequestHandler(types_js.ListPromptsRequestSchema, async (_request, extra) => {
        this.logger.debug("Handling ListPrompts request");
        if (this.definedPrompts) {
          return {
            prompts: this.definedPrompts?.map((p) => ({ ...p, version: p.version ?? void 0 }))
          };
        } else {
          try {
            const prompts = await capturedPromptOptions.listPrompts({ extra });
            for (const prompt of prompts) {
              types_js.PromptSchema.parse(prompt);
            }
            this.definedPrompts = prompts;
            this.logger.debug(`Fetched and cached ${this.definedPrompts.length} prompts.`);
            return {
              prompts: this.definedPrompts?.map((p) => ({ ...p, version: p.version ?? void 0 }))
            };
          } catch (error) {
            this.logger.error("Error fetching prompts via listPrompts():", {
              error: error instanceof Error ? error.message : String(error)
            });
            throw error;
          }
        }
      });
    }
    if (capturedPromptOptions.getPromptMessages) {
      serverInstance.setRequestHandler(
        types_js.GetPromptRequestSchema,
        async (request, extra) => {
          const startTime = Date.now();
          const { name, version, arguments: args } = request.params;
          if (!this.definedPrompts) {
            const prompts = await this.promptOptions?.listPrompts?.({ extra });
            if (!prompts) throw new Error("Failed to load prompts");
            this.definedPrompts = prompts;
          }
          let prompt;
          if (version) {
            prompt = this.definedPrompts?.find((p) => p.name === name && p.version === version);
          } else {
            prompt = this.definedPrompts?.find((p) => p.name === name);
          }
          if (!prompt) throw new Error(`Prompt "${name}"${version ? ` (version ${version})` : ""} not found`);
          if (prompt.arguments) {
            for (const arg of prompt.arguments) {
              if (arg.required && (args?.[arg.name] === void 0 || args?.[arg.name] === null)) {
                throw new Error(`Missing required argument: ${arg.name}`);
              }
            }
          }
          try {
            let messages = [];
            if (capturedPromptOptions.getPromptMessages) {
              messages = await capturedPromptOptions.getPromptMessages({ name, version, args, extra });
            }
            const duration = Date.now() - startTime;
            this.logger.info(
              `Prompt '${name}'${version ? ` (version ${version})` : ""} retrieved successfully in ${duration}ms.`
            );
            return { prompt, messages };
          } catch (error) {
            const duration = Date.now() - startTime;
            this.logger.error(`Failed to get content for prompt '${name}' in ${duration}ms`, { error });
            throw error;
          }
        }
      );
    }
  }
  convertAgentsToTools(agentsConfig, definedConvertedTools) {
    const agentTools = {};
    if (!agentsConfig) {
      return agentTools;
    }
    for (const agentKey in agentsConfig) {
      const agent = agentsConfig[agentKey];
      if (!agent || !("generate" in agent)) {
        this.logger.warn(`Agent instance for '${agentKey}' is invalid or missing a generate function. Skipping.`);
        continue;
      }
      const agentDescription = agent.getDescription();
      if (!agentDescription) {
        throw new Error(
          `Agent '${agent.name}' (key: '${agentKey}') must have a non-empty description to be used in an MCPServer.`
        );
      }
      const agentToolName = `ask_${agentKey}`;
      if (definedConvertedTools?.[agentToolName] || agentTools[agentToolName]) {
        this.logger.warn(
          `Tool with name '${agentToolName}' already exists. Agent '${agentKey}' will not be added as a duplicate tool.`
        );
        continue;
      }
      const agentToolDefinition = tools.createTool({
        id: agentToolName,
        description: `Ask agent '${agent.name}' a question. Agent description: ${agentDescription}`,
        inputSchema: zod.z.object({
          message: zod.z.string().describe("The question or input for the agent.")
        }),
        execute: async (inputData, context) => {
          this.logger.debug(
            `Executing agent tool '${agentToolName}' for agent '${agent.name}' with message: "${inputData.message}"`
          );
          try {
            const proxiedContext = context?.requestContext || new requestContext.RequestContext();
            if (context?.mcp?.extra) {
              Object.entries(context.mcp.extra).forEach(([key, value]) => {
                proxiedContext.set(key, value);
              });
            }
            const response = await agent.generate(inputData.message, {
              ...context ?? {},
              requestContext: proxiedContext
            });
            return response;
          } catch (error) {
            this.logger.error(`Error executing agent tool '${agentToolName}' for agent '${agent.name}':`, error);
            throw error;
          }
        }
      });
      const options = {
        name: agentToolName,
        logger: this.logger,
        mastra: this.mastra,
        requestContext: new requestContext.RequestContext(),
        tracingContext: {},
        description: agentToolDefinition.description
      };
      const coreTool = utils.makeCoreTool(agentToolDefinition, options);
      agentTools[agentToolName] = {
        ...coreTool,
        id: agentToolName,
        mcp: {
          toolType: "agent"
        }
      };
      this.logger.info(`Registered agent '${agent.name}' (key: '${agentKey}') as tool: '${agentToolName}'`);
    }
    return agentTools;
  }
  convertWorkflowsToTools(workflowsConfig, definedConvertedTools) {
    const workflowTools = {};
    if (!workflowsConfig) {
      return workflowTools;
    }
    for (const workflowKey in workflowsConfig) {
      const workflow = workflowsConfig[workflowKey];
      if (!workflow || typeof workflow.createRun !== "function") {
        this.logger.warn(
          `Workflow instance for '${workflowKey}' is invalid or missing a createRun function. Skipping.`
        );
        continue;
      }
      const workflowDescription = workflow.description;
      if (!workflowDescription) {
        throw new Error(
          `Workflow '${workflow.id}' (key: '${workflowKey}') must have a non-empty description to be used in an MCPServer.`
        );
      }
      const workflowToolName = `run_${workflowKey}`;
      if (definedConvertedTools?.[workflowToolName] || workflowTools[workflowToolName]) {
        this.logger.warn(
          `Tool with name '${workflowToolName}' already exists. Workflow '${workflowKey}' will not be added as a duplicate tool.`
        );
        continue;
      }
      const workflowToolDefinition = tools.createTool({
        id: workflowToolName,
        description: `Run workflow '${workflowKey}'. Workflow description: ${workflowDescription}`,
        inputSchema: workflow.inputSchema,
        execute: async (inputData, context) => {
          this.logger.debug(
            `Executing workflow tool '${workflowToolName}' for workflow '${workflow.id}' with input:`,
            inputData
          );
          try {
            const proxiedContext = context?.requestContext || new requestContext.RequestContext();
            if (context?.mcp?.extra) {
              Object.entries(context.mcp.extra).forEach(([key, value]) => {
                proxiedContext.set(key, value);
              });
            }
            const run2 = await workflow.createRun({ runId: proxiedContext?.get("runId") });
            const response = await run2.start({
              inputData,
              requestContext: proxiedContext,
              tracingContext: context?.tracingContext
            });
            return response;
          } catch (error) {
            this.logger.error(
              `Error executing workflow tool '${workflowToolName}' for workflow '${workflow.id}':`,
              error
            );
            throw error;
          }
        }
      });
      const options = {
        name: workflowToolName,
        logger: this.logger,
        mastra: this.mastra,
        requestContext: new requestContext.RequestContext(),
        tracingContext: {},
        description: workflowToolDefinition.description
      };
      const coreTool = utils.makeCoreTool(workflowToolDefinition, options);
      workflowTools[workflowToolName] = {
        ...coreTool,
        id: workflowToolName,
        mcp: {
          toolType: "workflow"
        }
      };
      this.logger.info(`Registered workflow '${workflow.id}' (key: '${workflowKey}') as tool: '${workflowToolName}'`);
    }
    return workflowTools;
  }
  /**
   * Convert and validate all provided tools, logging registration status.
   * Also converts agents and workflows into tools.
   * @param tools Tool definitions
   * @param agentsConfig Agent definitions to be converted to tools, expected from MCPServerConfig
   * @param workflowsConfig Workflow definitions to be converted to tools, expected from MCPServerConfig
   * @returns Converted tools registry
   */
  convertTools(tools, agentsConfig, workflowsConfig) {
    const definedConvertedTools = {};
    for (const toolName of Object.keys(tools)) {
      const toolInstance = tools[toolName];
      if (!toolInstance) {
        this.logger.warn(`Tool instance for '${toolName}' is undefined. Skipping.`);
        continue;
      }
      if (typeof toolInstance.execute !== "function") {
        this.logger.warn(`Tool '${toolName}' does not have a valid execute function. Skipping.`);
        continue;
      }
      const options = {
        name: toolName,
        requestContext: new requestContext.RequestContext(),
        tracingContext: {},
        mastra: this.mastra,
        logger: this.logger,
        description: toolInstance?.description
      };
      const coreTool = utils.makeCoreTool(toolInstance, options);
      definedConvertedTools[toolName] = {
        ...coreTool,
        id: toolName
      };
      this.logger.info(`Registered explicit tool: '${toolName}'`);
    }
    this.logger.info(`Total defined tools registered: ${Object.keys(definedConvertedTools).length}`);
    let agentDerivedTools = {};
    let workflowDerivedTools = {};
    try {
      agentDerivedTools = this.convertAgentsToTools(agentsConfig, definedConvertedTools);
      workflowDerivedTools = this.convertWorkflowsToTools(workflowsConfig, definedConvertedTools);
    } catch (e) {
      const mastraError = new error.MastraError(
        {
          id: "MCP_SERVER_AGENT_OR_WORKFLOW_TOOL_CONVERSION_FAILED",
          domain: error.ErrorDomain.MCP,
          category: error.ErrorCategory.USER
        },
        e
      );
      this.logger.trackException(mastraError);
      this.logger.error("Failed to convert tools:", {
        error: mastraError.toString()
      });
      throw mastraError;
    }
    const allConvertedTools = { ...definedConvertedTools, ...agentDerivedTools, ...workflowDerivedTools };
    const finalToolCount = Object.keys(allConvertedTools).length;
    const definedCount = Object.keys(definedConvertedTools).length;
    const fromAgentsCount = Object.keys(agentDerivedTools).length;
    const fromWorkflowsCount = Object.keys(workflowDerivedTools).length;
    this.logger.info(
      `${finalToolCount} total tools registered (${definedCount} defined + ${fromAgentsCount} agents + ${fromWorkflowsCount} workflows)`
    );
    return allConvertedTools;
  }
  /**
   * Starts the MCP server using standard input/output (stdio) transport.
   *
   * This is typically used when running the server as a command-line program that MCP clients
   * spawn as a subprocess (e.g., integration with Windsurf, Cursor, or Claude Desktop).
   *
   * @throws {MastraError} If the stdio connection fails
   *
   * @example
   * ```typescript
   * const server = new MCPServer({
   *   name: 'My Server',
   *   version: '1.0.0',
   *   tools: { weatherTool },
   * });
   *
   * await server.startStdio();
   * ```
   */
  async startStdio() {
    this.stdioTransport = new stdio_js$1.StdioServerTransport();
    try {
      await this.server.connect(this.stdioTransport);
    } catch (error$1) {
      const mastraError = new error.MastraError(
        {
          id: "MCP_SERVER_STDIO_CONNECTION_FAILED",
          domain: error.ErrorDomain.MCP,
          category: error.ErrorCategory.THIRD_PARTY
        },
        error$1
      );
      this.logger.trackException(mastraError);
      this.logger.error("Failed to connect MCP server using stdio transport:", {
        error: mastraError.toString()
      });
      throw mastraError;
    }
    this.logger.info("Started MCP Server (stdio)");
  }
  /**
   * Integrates the MCP server with an existing HTTP server using Server-Sent Events (SSE).
   *
   * Call this method from your web server's request handler for both the SSE and message paths.
   * This enables web-based MCP clients to connect to your server.
   *
   * @param options - Configuration for SSE integration
   * @param options.url - Parsed URL of the incoming request
   * @param options.ssePath - Path for establishing SSE connection (e.g., '/sse')
   * @param options.messagePath - Path for POSTing client messages (e.g., '/message')
   * @param options.req - Incoming HTTP request object
   * @param options.res - HTTP response object (must support .write/.end)
   *
   * @throws {MastraError} If SSE connection setup fails
   *
   * @example
   * ```typescript
   * import http from 'node:http';
   *
   * const httpServer = http.createServer(async (req, res) => {
   *   await server.startSSE({
   *     url: new URL(req.url || '', `http://localhost:1234`),
   *     ssePath: '/sse',
   *     messagePath: '/message',
   *     req,
   *     res,
   *   });
   * });
   *
   * httpServer.listen(1234, () => {
   *   console.log('MCP server listening on http://localhost:1234/sse');
   * });
   * ```
   */
  async startSSE({ url, ssePath, messagePath, req, res }) {
    try {
      if (url.pathname === ssePath) {
        await this.connectSSE({
          messagePath,
          res
        });
      } else if (url.pathname === messagePath) {
        this.logger.debug("Received message");
        if (!this.sseTransport) {
          res.writeHead(503);
          res.end("SSE connection not established");
          return;
        }
        const parsedBody = await this.readJsonBody(req, { preParsedOnly: true });
        await this.sseTransport.handlePostMessage(req, res, parsedBody);
      } else {
        this.logger.debug("Unknown path:", { path: url.pathname });
        res.writeHead(404);
        res.end();
      }
    } catch (e) {
      const mastraError = new error.MastraError(
        {
          id: "MCP_SERVER_SSE_START_FAILED",
          domain: error.ErrorDomain.MCP,
          category: error.ErrorCategory.USER,
          details: {
            url: url.toString(),
            ssePath,
            messagePath
          }
        },
        e
      );
      this.logger.trackException(mastraError);
      this.logger.error("Failed to start MCP Server (SSE):", { error: mastraError.toString() });
      throw mastraError;
    }
  }
  /**
   * Integrates the MCP server with a Hono web framework using Server-Sent Events (SSE).
   *
   * Call this method from your Hono server's request handler for both the SSE and message paths.
   * This enables Hono-based web applications to expose MCP servers.
   *
   * @param options - Configuration for Hono SSE integration
   * @param options.url - Parsed URL of the incoming request
   * @param options.ssePath - Path for establishing SSE connection (e.g., '/hono-sse')
   * @param options.messagePath - Path for POSTing client messages (e.g., '/message')
   * @param options.context - Hono context object
   *
   * @throws {MastraError} If Hono SSE connection setup fails
   *
   * @example
   * ```typescript
   * import { Hono } from 'hono';
   *
   * const app = new Hono();
   *
   * app.all('*', async (c) => {
   *   const url = new URL(c.req.url);
   *   return await server.startHonoSSE({
   *     url,
   *     ssePath: '/hono-sse',
   *     messagePath: '/message',
   *     context: c,
   *   });
   * });
   *
   * export default app;
   * ```
   */
  async startHonoSSE({ url, ssePath, messagePath, context }) {
    try {
      if (url.pathname === ssePath) {
        return streamSSE(context, async (stream2) => {
          await this.connectHonoSSE({
            messagePath,
            stream: stream2
          });
        });
      } else if (url.pathname === messagePath) {
        this.logger.debug("Received message");
        const sessionId = context.req.query("sessionId");
        this.logger.debug("Received message for sessionId", { sessionId });
        if (!sessionId) {
          return context.text("No sessionId provided", 400);
        }
        if (!this.sseHonoTransports.has(sessionId)) {
          return context.text(`No transport found for sessionId ${sessionId}`, 400);
        }
        const message = await this.sseHonoTransports.get(sessionId)?.handlePostMessage(context);
        if (!message) {
          return context.text("Transport not found", 400);
        }
        return message;
      } else {
        this.logger.debug("Unknown path:", { path: url.pathname });
        return context.text("Unknown path", 404);
      }
    } catch (e) {
      const mastraError = new error.MastraError(
        {
          id: "MCP_SERVER_HONO_SSE_START_FAILED",
          domain: error.ErrorDomain.MCP,
          category: error.ErrorCategory.USER,
          details: {
            url: url.toString(),
            ssePath,
            messagePath
          }
        },
        e
      );
      this.logger.trackException(mastraError);
      this.logger.error("Failed to start MCP Server (Hono SSE):", { error: mastraError.toString() });
      throw mastraError;
    }
  }
  /**
   * Integrates the MCP server with an existing HTTP server using streamable HTTP transport.
   *
   * This is the recommended modern transport method, providing better session management and
   * reliability compared to SSE. Call this from your HTTP server's request handler.
   *
   * @param options - Configuration for HTTP integration
   * @param options.url - Parsed URL of the incoming request
   * @param options.httpPath - Path for the MCP endpoint (e.g., '/mcp')
   * @param options.req - Incoming HTTP request (http.IncomingMessage)
   * @param options.res - HTTP response object (http.ServerResponse)
   * @param options.options - Optional transport options
   * @param options.options.sessionIdGenerator - Function to generate unique session IDs (defaults to randomUUID)
   * @param options.options.onsessioninitialized - Callback when a new session is initialized
   * @param options.options.enableJsonResponse - If true, return JSON instead of SSE streaming
   * @param options.options.eventStore - Event store for message resumability
   * @param options.options.serverless - If true, run in stateless mode without session management (ideal for serverless environments)
   *
   * @throws {MastraError} If HTTP connection setup fails
   *
   * @example
   * ```typescript
   * import http from 'node:http';
   * import { randomUUID } from 'node:crypto';
   *
   * const httpServer = http.createServer(async (req, res) => {
   *   await server.startHTTP({
   *     url: new URL(req.url || '', 'http://localhost:1234'),
   *     httpPath: '/mcp',
   *     req,
   *     res,
   *     options: {
   *       sessionIdGenerator: () => randomUUID(),
   *       onsessioninitialized: (sessionId) => {
   *         console.log(`New MCP session: ${sessionId}`);
   *       },
   *     },
   *   });
   * });
   *
   * httpServer.listen(1234);
   * ```
   *
   * @example Serverless mode (Cloudflare Workers, Vercel Edge, etc.)
   * ```typescript
   * export default {
   *   async fetch(request: Request) {
   *     const url = new URL(request.url);
   *     if (url.pathname === '/mcp') {
   *       await server.startHTTP({
   *         url,
   *         httpPath: '/mcp',
   *         req: request,
   *         res: response,
   *         options: { serverless: true },
   *       });
   *     }
   *     return new Response('Not found', { status: 404 });
   *   },
   * };
   * ```
   */
  async startHTTP({
    url,
    httpPath,
    req,
    res,
    options
  }) {
    this.logger.debug(`startHTTP: Received ${req.method} request to ${url.pathname}`);
    if (url.pathname !== httpPath) {
      this.logger.debug(`startHTTP: Pathname ${url.pathname} does not match httpPath ${httpPath}. Returning 404.`);
      res.writeHead(404);
      res.end();
      return;
    }
    const isStatelessMode = options?.serverless || options && "sessionIdGenerator" in options && options.sessionIdGenerator === void 0;
    if (isStatelessMode) {
      this.logger.debug("startHTTP: Running in stateless mode (serverless or sessionIdGenerator: undefined)");
      await this.handleServerlessRequest(req, res);
      return;
    }
    const mergedOptions = {
      sessionIdGenerator: () => crypto$1.randomUUID(),
      // default: enabled
      ...options
      // user-provided overrides default
    };
    const sessionId = req.headers["mcp-session-id"];
    let transport;
    this.logger.debug(
      `startHTTP: Session ID from headers: ${sessionId}. Active transports: ${Array.from(this.streamableHTTPTransports.keys()).join(", ")}`
    );
    try {
      if (sessionId && this.streamableHTTPTransports.has(sessionId)) {
        transport = this.streamableHTTPTransports.get(sessionId);
        this.logger.debug(`startHTTP: Using existing Streamable HTTP transport for session ID: ${sessionId}`);
        if (req.method === "GET") {
          this.logger.debug(
            `startHTTP: Handling GET request for existing session ${sessionId}. Calling transport.handleRequest.`
          );
        }
        const body = req.method === "POST" ? await this.readJsonBody(req) : void 0;
        await transport.handleRequest(req, res, body);
      } else {
        this.logger.debug(`startHTTP: No existing Streamable HTTP session ID found. ${req.method}`);
        if (req.method === "POST") {
          const body = await this.readJsonBody(req);
          const { isInitializeRequest } = await import('@modelcontextprotocol/sdk/types.js');
          if (isInitializeRequest(body)) {
            this.logger.debug("startHTTP: Received Streamable HTTP initialize request, creating new transport.");
            transport = new streamableHttp_js$1.StreamableHTTPServerTransport({
              ...mergedOptions,
              sessionIdGenerator: mergedOptions.sessionIdGenerator,
              onsessioninitialized: (id) => {
                this.streamableHTTPTransports.set(id, transport);
              }
            });
            transport.onclose = () => {
              const closedSessionId = transport?.sessionId;
              if (closedSessionId && this.streamableHTTPTransports.has(closedSessionId)) {
                this.logger.debug(
                  `startHTTP: Streamable HTTP transport closed for session ${closedSessionId}, removing from map.`
                );
                this.streamableHTTPTransports.delete(closedSessionId);
                if (this.httpServerInstances.has(closedSessionId)) {
                  this.httpServerInstances.delete(closedSessionId);
                  this.logger.debug(`startHTTP: Cleaned up server instance for closed session ${closedSessionId}`);
                }
              }
            };
            const sessionServerInstance = this.createServerInstance();
            await sessionServerInstance.connect(transport);
            if (transport.sessionId) {
              this.streamableHTTPTransports.set(transport.sessionId, transport);
              this.httpServerInstances.set(transport.sessionId, sessionServerInstance);
              this.logger.debug(
                `startHTTP: Streamable HTTP session initialized and stored with ID: ${transport.sessionId}`
              );
            } else {
              this.logger.warn("startHTTP: Streamable HTTP transport initialized without a session ID.");
            }
            return await transport.handleRequest(req, res, body);
          } else {
            this.logger.warn("startHTTP: Received non-initialize POST request without a session ID.");
            res.writeHead(400, { "Content-Type": "application/json" });
            res.end(
              JSON.stringify({
                jsonrpc: "2.0",
                error: {
                  code: -32e3,
                  message: "Bad Request: No valid session ID provided for non-initialize request"
                },
                id: body?.id ?? null
                // Include original request ID if available
              })
            );
          }
        } else {
          this.logger.warn(`startHTTP: Received ${req.method} request without a session ID.`);
          res.writeHead(400, { "Content-Type": "application/json" });
          res.end(
            JSON.stringify({
              jsonrpc: "2.0",
              error: {
                code: -32e3,
                message: `Bad Request: ${req.method} request requires a valid session ID`
              },
              id: null
            })
          );
        }
      }
    } catch (error$1) {
      const mastraError = new error.MastraError(
        {
          id: "MCP_SERVER_HTTP_CONNECTION_FAILED",
          domain: error.ErrorDomain.MCP,
          category: error.ErrorCategory.USER,
          text: "Failed to connect MCP server using HTTP transport"
        },
        error$1
      );
      this.logger.trackException(mastraError);
      this.logger.error("startHTTP: Error handling Streamable HTTP request:", { error: mastraError });
      if (!res.headersSent) {
        res.writeHead(500, { "Content-Type": "application/json" });
        res.end(
          JSON.stringify({
            jsonrpc: "2.0",
            error: {
              code: -32603,
              message: "Internal server error"
            },
            id: null
            // Cannot determine original request ID in catch
          })
        );
      }
    }
  }
  /**
   * Handles a stateless, serverless HTTP request without session management.
   *
   * This method bypasses all session/transport state and handles each request independently.
   * For serverless environments (Cloudflare Workers, Vercel Edge, etc.) where
   * persistent connections and session state cannot be maintained across requests.
   *
   * Each request gets a fresh transport and server instance that are discarded after the response.
   *
   * @param req - Incoming HTTP request
   * @param res - HTTP response object
   * @private
   */
  async handleServerlessRequest(req, res) {
    try {
      this.logger.debug(`handleServerlessRequest: Received ${req.method} request`);
      const body = req.method === "POST" ? await this.readJsonBody(req) : void 0;
      this.logger.debug(`handleServerlessRequest: Processing ${req.method} request`, {
        method: body?.method,
        id: body?.id
      });
      const transientServer = this.createServerInstance();
      const tempTransport = new streamableHttp_js$1.StreamableHTTPServerTransport({
        sessionIdGenerator: void 0,
        enableJsonResponse: true
      });
      await transientServer.connect(tempTransport);
      await tempTransport.handleRequest(req, res, body);
      this.logger.debug(`handleServerlessRequest: Completed ${body?.method} request`, { id: body?.id });
    } catch (error$1) {
      const mastraError = new error.MastraError(
        {
          id: "MCP_SERVER_SERVERLESS_REQUEST_FAILED",
          domain: error.ErrorDomain.MCP,
          category: error.ErrorCategory.USER,
          text: "Failed to handle serverless MCP request"
        },
        error$1
      );
      this.logger.trackException(mastraError);
      this.logger.error("handleServerlessRequest: Error handling request:", { error: mastraError });
      if (!res.headersSent) {
        res.writeHead(500, { "Content-Type": "application/json" });
        res.end(
          JSON.stringify({
            jsonrpc: "2.0",
            error: {
              code: -32603,
              message: "Internal server error",
              data: error$1 instanceof Error ? error$1.message : String(error$1)
            },
            id: null
          })
        );
      }
    }
  }
  /**
   * Establishes the SSE connection for the MCP server.
   *
   * This is a lower-level method called internally by `startSSE()`. In most cases,
   * you should use `startSSE()` instead which handles both connection establishment
   * and message routing.
   *
   * @param params - Connection parameters
   * @param params.messagePath - Path for POST requests from the client
   * @param params.res - HTTP response object for the SSE stream
   * @throws {MastraError} If SSE connection establishment fails
   *
   * @example
   * ```typescript
   * // Usually called internally by startSSE()
   * await server.connectSSE({
   *   messagePath: '/message',
   *   res: response
   * });
   * ```
   */
  async connectSSE({
    messagePath,
    res
  }) {
    try {
      this.logger.debug("Received SSE connection");
      this.sseTransport = new sse_js$1.SSEServerTransport(messagePath, res);
      await this.server.connect(this.sseTransport);
      this.server.onclose = async () => {
        this.sseTransport = void 0;
        await this.server.close();
      };
      res.on("close", () => {
        this.sseTransport = void 0;
      });
    } catch (e) {
      const mastraError = new error.MastraError(
        {
          id: "MCP_SERVER_SSE_CONNECT_FAILED",
          domain: error.ErrorDomain.MCP,
          category: error.ErrorCategory.USER,
          details: {
            messagePath
          }
        },
        e
      );
      this.logger.trackException(mastraError);
      this.logger.error("Failed to connect to MCP Server (SSE):", { error: mastraError });
      throw mastraError;
    }
  }
  /**
   * Establishes the Hono SSE connection for the MCP server.
   *
   * This is a lower-level method called internally by `startHonoSSE()`. In most cases,
   * you should use `startHonoSSE()` instead which handles both connection establishment
   * and message routing.
   *
   * @param params - Connection parameters
   * @param params.messagePath - Path for POST requests from the client
   * @param params.stream - Hono SSE streaming API object
   * @throws {MastraError} If Hono SSE connection establishment fails
   *
   * @example
   * ```typescript
   * // Usually called internally by startHonoSSE()
   * await server.connectHonoSSE({
   *   messagePath: '/message',
   *   stream: sseStream
   * });
   * ```
   */
  async connectHonoSSE({ messagePath, stream: stream2 }) {
    this.logger.debug("Received SSE connection");
    const sseTransport = new SSETransport(messagePath, stream2);
    const sessionId = sseTransport.sessionId;
    this.logger.debug("SSE Transport created with sessionId:", { sessionId });
    this.sseHonoTransports.set(sessionId, sseTransport);
    stream2.onAbort(() => {
      this.logger.debug("SSE Transport aborted with sessionId:", { sessionId });
      this.sseHonoTransports.delete(sessionId);
    });
    try {
      await this.server.connect(sseTransport);
      this.server.onclose = async () => {
        this.logger.debug("SSE Transport closed with sessionId:", { sessionId });
        this.sseHonoTransports.delete(sessionId);
        await this.server.close();
      };
      while (true) {
        await stream2.sleep(6e4);
        const sessionIds = Array.from(this.sseHonoTransports.keys() || []);
        this.logger.debug("Active Hono SSE sessions:", { sessionIds });
        await stream2.write(":keep-alive\n\n");
      }
    } catch (e) {
      const mastraError = new error.MastraError(
        {
          id: "MCP_SERVER_HONO_SSE_CONNECT_FAILED",
          domain: error.ErrorDomain.MCP,
          category: error.ErrorCategory.USER,
          details: {
            messagePath
          }
        },
        e
      );
      this.logger.trackException(mastraError);
      this.logger.error("Failed to connect to MCP Server (Hono SSE):", { error: mastraError });
      throw mastraError;
    }
  }
  /**
   * Closes the MCP server and releases all resources.
   *
   * This method cleanly shuts down all active transports (stdio, SSE, HTTP) and their
   * associated connections. Call this when your application is shutting down.
   *
   * @throws {MastraError} If closing the server fails
   *
   * @example
   * ```typescript
   * // Graceful shutdown
   * process.on('SIGTERM', async () => {
   *   await server.close();
   *   process.exit(0);
   * });
   * ```
   */
  async close() {
    try {
      if (this.stdioTransport) {
        await this.stdioTransport.close?.();
        this.stdioTransport = void 0;
      }
      if (this.sseTransport) {
        await this.sseTransport.close?.();
        this.sseTransport = void 0;
      }
      if (this.sseHonoTransports) {
        for (const transport of this.sseHonoTransports.values()) {
          await transport.close?.();
        }
        this.sseHonoTransports.clear();
      }
      if (this.streamableHTTPTransports) {
        for (const transport of this.streamableHTTPTransports.values()) {
          await transport.close?.();
        }
        this.streamableHTTPTransports.clear();
      }
      if (this.httpServerInstances) {
        for (const serverInstance of this.httpServerInstances.values()) {
          await serverInstance.close?.();
        }
        this.httpServerInstances.clear();
      }
      await this.server.close();
      this.logger.info("MCP server closed.");
    } catch (error$1) {
      const mastraError = new error.MastraError(
        {
          id: "MCP_SERVER_CLOSE_FAILED",
          domain: error.ErrorDomain.MCP,
          category: error.ErrorCategory.THIRD_PARTY
        },
        error$1
      );
      this.logger.trackException(mastraError);
      this.logger.error("Error closing MCP server:", { error: mastraError });
      throw mastraError;
    }
  }
  /**
   * Gets basic information about the server.
   *
   * Returns metadata including server ID, name, description, repository, and version details.
   * This information conforms to the MCP Server schema.
   *
   * @returns Server information object
   *
   * @example
   * ```typescript
   * const info = server.getServerInfo();
   * console.log(`${info.name} v${info.version_detail.version}`);
   * // Output: My Weather Server v1.0.0
   * ```
   */
  getServerInfo() {
    return {
      id: this.id,
      name: this.name,
      description: this.description,
      repository: this.repository,
      version_detail: {
        version: this.version,
        release_date: this.releaseDate,
        is_latest: this.isLatest
      }
    };
  }
  /**
   * Gets detailed information about the server including packaging and deployment metadata.
   *
   * Returns extended server information with package details, remotes, and deployment configurations.
   * This information conforms to the MCP ServerDetail schema.
   *
   * @returns Detailed server information object
   *
   * @example
   * ```typescript
   * const detail = server.getServerDetail();
   * console.log(detail.package_canonical); // 'npm'
   * console.log(detail.packages); // Package installation info
   * ```
   */
  getServerDetail() {
    return {
      ...this.getServerInfo(),
      package_canonical: this.packageCanonical,
      packages: this.packages,
      remotes: this.remotes
    };
  }
  /**
   * Gets a list of all tools provided by this MCP server with their schemas.
   *
   * Returns information about all registered tools including explicit tools, agent-derived tools,
   * and workflow-derived tools. Includes input/output schemas and tool types.
   *
   * @returns Object containing array of tool information
   *
   * @example
   * ```typescript
   * const toolList = server.getToolListInfo();
   * toolList.tools.forEach(tool => {
   *   console.log(`${tool.name}: ${tool.description}`);
   *   console.log(`Type: ${tool.toolType || 'tool'}`);
   * });
   * ```
   */
  getToolListInfo() {
    this.logger.debug(`Getting tool list information for MCPServer '${this.name}'`);
    return {
      tools: Object.entries(this.convertedTools).map(([toolId, tool]) => ({
        id: toolId,
        name: tool.id || toolId,
        description: tool.description,
        inputSchema: tool.parameters?.jsonSchema || tool.parameters,
        outputSchema: tool.outputSchema?.jsonSchema || tool.outputSchema,
        toolType: tool.mcp?.toolType
      }))
    };
  }
  /**
   * Gets information for a specific tool provided by this MCP server.
   *
   * Returns detailed information about a single tool including its name, description, schemas, and type.
   * Returns undefined if the tool is not found.
   *
   * @param toolId - The ID/name of the tool to retrieve
   * @returns Tool information object or undefined if not found
   *
   * @example
   * ```typescript
   * const toolInfo = server.getToolInfo('getWeather');
   * if (toolInfo) {
   *   console.log(toolInfo.description);
   *   console.log(toolInfo.inputSchema);
   * }
   * ```
   */
  getToolInfo(toolId) {
    const tool = this.convertedTools[toolId];
    if (!tool) {
      this.logger.debug(`Tool '${toolId}' not found on MCPServer '${this.name}'`);
      return void 0;
    }
    this.logger.debug(`Getting info for tool '${toolId}' on MCPServer '${this.name}'`);
    return {
      name: tool.id || toolId,
      description: tool.description,
      inputSchema: tool.parameters?.jsonSchema || tool.parameters,
      outputSchema: tool.outputSchema?.jsonSchema || tool.outputSchema,
      toolType: tool.mcp?.toolType
    };
  }
  /**
   * Executes a specific tool provided by this MCP server.
   *
   * This method validates the tool arguments against the input schema and executes the tool.
   * If validation fails, returns an error object instead of throwing.
   *
   * @param toolId - The ID/name of the tool to execute
   * @param args - The arguments to pass to the tool's execute function
   * @param executionContext - Optional context including messages and toolCallId
   * @returns Promise resolving to the tool execution result
   * @throws {MastraError} If the tool is not found or execution fails
   *
   * @example
   * ```typescript
   * const result = await server.executeTool(
   *   'getWeather',
   *   { location: 'London' },
   *   { toolCallId: 'call_123' }
   * );
   * console.log(result);
   * ```
   */
  async executeTool(toolId, args, executionContext) {
    const tool = this.convertedTools[toolId];
    let validatedArgs = args;
    try {
      if (!tool) {
        this.logger.warn(`ExecuteTool: Unknown tool '${toolId}' requested on MCPServer '${this.name}'.`);
        throw new Error(`Unknown tool: ${toolId}`);
      }
      this.logger.debug(`ExecuteTool: Invoking '${toolId}' with arguments:`, args);
      if (tool.parameters instanceof zod.z.ZodType && typeof tool.parameters.safeParse === "function") {
        const validation = tool.parameters.safeParse(args ?? {});
        if (!validation.success) {
          const errorMessages = validation.error.errors.map((e) => `- ${e.path?.join(".") || "root"}: ${e.message}`).join("\n");
          this.logger.warn(`ExecuteTool: Invalid tool arguments for '${toolId}': ${errorMessages}`, {
            errors: validation.error.format()
          });
          return {
            error: true,
            message: `Tool validation failed. Please fix the following errors and try again:
${errorMessages}

Provided arguments: ${JSON.stringify(args, null, 2)}`,
            validationErrors: validation.error.format()
          };
        }
        validatedArgs = validation.data;
      } else {
        this.logger.debug(
          `ExecuteTool: Tool '${toolId}' parameters is not a Zod schema with safeParse or is undefined. Skipping validation.`
        );
      }
      if (!tool.execute) {
        this.logger.error(`ExecuteTool: Tool '${toolId}' does not have an execute function.`);
        throw new Error(`Tool '${toolId}' cannot be executed.`);
      }
    } catch (error$1) {
      const mastraError = new error.MastraError(
        {
          id: "MCP_SERVER_TOOL_EXECUTE_PREPARATION_FAILED",
          domain: error.ErrorDomain.MCP,
          category: error.ErrorCategory.USER,
          details: {
            toolId,
            args
          }
        },
        error$1
      );
      this.logger.trackException(mastraError);
      throw mastraError;
    }
    try {
      const finalExecutionContext = {
        messages: executionContext?.messages || [],
        toolCallId: executionContext?.toolCallId || crypto$1.randomUUID()
      };
      const result = await tool.execute(validatedArgs, finalExecutionContext);
      this.logger.info(`ExecuteTool: Tool '${toolId}' executed successfully.`);
      return result;
    } catch (error$1) {
      const mastraError = new error.MastraError(
        {
          id: "MCP_SERVER_TOOL_EXECUTE_FAILED",
          domain: error.ErrorDomain.MCP,
          category: error.ErrorCategory.USER,
          details: {
            toolId,
            validatedArgs
          }
        },
        error$1
      );
      this.logger.trackException(mastraError);
      this.logger.error(`ExecuteTool: Tool execution failed for '${toolId}':`, { error: error$1 });
      throw mastraError;
    }
  }
};
function escapeHeaderValue(value) {
  return value.replace(/\\/g, "\\\\").replace(/"/g, '\\"');
}
function generateWWWAuthenticateHeader(options = {}) {
  const params = [];
  if (options.resourceMetadataUrl) {
    params.push(`resource_metadata="${escapeHeaderValue(options.resourceMetadataUrl)}"`);
  }
  if (options.additionalParams) {
    for (const [key, value] of Object.entries(options.additionalParams)) {
      params.push(`${key}="${escapeHeaderValue(value)}"`);
    }
  }
  if (params.length === 0) {
    return "Bearer";
  }
  return `Bearer ${params.join(", ")}`;
}
function generateProtectedResourceMetadata(config) {
  return {
    resource: config.resource,
    authorization_servers: config.authorizationServers,
    scopes_supported: config.scopesSupported ?? ["mcp:read", "mcp:write"],
    bearer_methods_supported: ["header"],
    ...config.resourceName && { resource_name: config.resourceName },
    ...config.resourceDocumentation && {
      resource_documentation: config.resourceDocumentation
    }
  };
}
function extractBearerToken(authHeader) {
  if (!authHeader) return void 0;
  const prefix = "bearer ";
  if (authHeader.length <= prefix.length) return void 0;
  if (authHeader.slice(0, prefix.length).toLowerCase() !== prefix) return void 0;
  const token = authHeader.slice(prefix.length).trim();
  return token || void 0;
}

// src/server/oauth-middleware.ts
function createOAuthMiddleware(options) {
  const { oauth, mcpPath = "/mcp", logger } = options;
  const protectedResourceMetadata = generateProtectedResourceMetadata(oauth);
  const wellKnownPath = "/.well-known/oauth-protected-resource";
  const resourceMetadataUrl = new URL(wellKnownPath, oauth.resource).toString();
  return async function oauthMiddleware(req, res, url) {
    logger?.debug?.(`OAuth middleware: ${req.method} ${url.pathname}`);
    if (url.pathname === wellKnownPath && req.method === "GET") {
      logger?.debug?.("OAuth middleware: Serving Protected Resource Metadata");
      res.writeHead(200, {
        "Content-Type": "application/json",
        "Cache-Control": "max-age=3600",
        "Access-Control-Allow-Origin": "*"
      });
      res.end(JSON.stringify(protectedResourceMetadata));
      return { proceed: false, handled: true };
    }
    if (url.pathname === wellKnownPath && req.method === "OPTIONS") {
      res.writeHead(204, {
        "Access-Control-Allow-Origin": "*",
        "Access-Control-Allow-Methods": "GET, OPTIONS",
        "Access-Control-Allow-Headers": "Content-Type",
        "Access-Control-Max-Age": "86400"
      });
      res.end();
      return { proceed: false, handled: true };
    }
    if (!url.pathname.startsWith(mcpPath)) {
      return { proceed: true, handled: false };
    }
    const authHeader = req.headers["authorization"];
    const token = extractBearerToken(authHeader);
    if (!token) {
      logger?.debug?.("OAuth middleware: No bearer token provided");
      res.writeHead(401, {
        "Content-Type": "application/json",
        "WWW-Authenticate": generateWWWAuthenticateHeader({ resourceMetadataUrl })
      });
      res.end(
        JSON.stringify({
          error: "unauthorized",
          error_description: "Bearer token required"
        })
      );
      return { proceed: false, handled: true };
    }
    if (oauth.validateToken) {
      logger?.debug?.("OAuth middleware: Validating token");
      const validationResult = await oauth.validateToken(token, oauth.resource);
      if (!validationResult.valid) {
        logger?.debug?.(`OAuth middleware: Token validation failed: ${validationResult.error}`);
        res.writeHead(401, {
          "Content-Type": "application/json",
          "WWW-Authenticate": generateWWWAuthenticateHeader({
            resourceMetadataUrl,
            additionalParams: {
              error: validationResult.error || "invalid_token",
              ...validationResult.errorDescription && {
                error_description: validationResult.errorDescription
              }
            }
          })
        });
        res.end(
          JSON.stringify({
            error: validationResult.error || "invalid_token",
            error_description: validationResult.errorDescription || "Token validation failed"
          })
        );
        return { proceed: false, handled: true, tokenValidation: validationResult };
      }
      logger?.debug?.("OAuth middleware: Token validated successfully");
      return { proceed: true, handled: false, tokenValidation: validationResult };
    }
    logger?.debug?.("OAuth middleware: No token validation configured, accepting token");
    return {
      proceed: true,
      handled: false,
      tokenValidation: { valid: true }
    };
  };
}
function createStaticTokenValidator(validTokens) {
  const tokenSet = new Set(validTokens);
  return async (token) => {
    if (tokenSet.has(token)) {
      return { valid: true, scopes: ["mcp:read", "mcp:write"] };
    }
    return {
      valid: false,
      error: "invalid_token",
      errorDescription: "Token not recognized"
    };
  };
}
function createIntrospectionValidator(introspectionEndpoint, clientCredentials) {
  return async (token, resource) => {
    try {
      const headers = {
        "Content-Type": "application/x-www-form-urlencoded"
      };
      if (clientCredentials) {
        if (clientCredentials.clientId.includes(":")) {
          return {
            valid: false,
            error: "invalid_request",
            errorDescription: "clientId cannot contain a colon character per RFC 7617"
          };
        }
        const credentials = Buffer.from(`${clientCredentials.clientId}:${clientCredentials.clientSecret}`).toString(
          "base64"
        );
        headers["Authorization"] = `Basic ${credentials}`;
      }
      const response = await fetch(introspectionEndpoint, {
        method: "POST",
        headers,
        body: new URLSearchParams({
          token,
          token_type_hint: "access_token"
        })
      });
      if (!response.ok) {
        return {
          valid: false,
          error: "server_error",
          errorDescription: `Introspection failed: ${response.status}`
        };
      }
      const data = await response.json();
      if (!data.active) {
        return {
          valid: false,
          error: "invalid_token",
          errorDescription: "Token is not active"
        };
      }
      if (data.aud) {
        const audiences = Array.isArray(data.aud) ? data.aud : [data.aud];
        if (!audiences.includes(resource)) {
          return {
            valid: false,
            error: "invalid_token",
            errorDescription: "Token audience does not match this resource"
          };
        }
      }
      return {
        valid: true,
        scopes: data.scope?.trim().split(" ").filter((s) => s !== "") || [],
        subject: data.sub,
        expiresAt: data.exp,
        claims: data
      };
    } catch (error) {
      return {
        valid: false,
        error: "server_error",
        errorDescription: error instanceof Error ? error.message : "Introspection failed"
      };
    }
  };
}

Object.defineProperty(exports, "UnauthorizedError", {
  enumerable: true,
  get: function () { return auth_js.UnauthorizedError; }
});
Object.defineProperty(exports, "auth", {
  enumerable: true,
  get: function () { return auth_js.auth; }
});
Object.defineProperty(exports, "buildDiscoveryUrls", {
  enumerable: true,
  get: function () { return auth_js.buildDiscoveryUrls; }
});
Object.defineProperty(exports, "discoverAuthorizationServerMetadata", {
  enumerable: true,
  get: function () { return auth_js.discoverAuthorizationServerMetadata; }
});
Object.defineProperty(exports, "discoverOAuthMetadata", {
  enumerable: true,
  get: function () { return auth_js.discoverOAuthMetadata; }
});
Object.defineProperty(exports, "discoverOAuthProtectedResourceMetadata", {
  enumerable: true,
  get: function () { return auth_js.discoverOAuthProtectedResourceMetadata; }
});
Object.defineProperty(exports, "exchangeAuthorization", {
  enumerable: true,
  get: function () { return auth_js.exchangeAuthorization; }
});
Object.defineProperty(exports, "extractResourceMetadataUrl", {
  enumerable: true,
  get: function () { return auth_js.extractResourceMetadataUrl; }
});
Object.defineProperty(exports, "parseErrorResponse", {
  enumerable: true,
  get: function () { return auth_js.parseErrorResponse; }
});
Object.defineProperty(exports, "refreshAuthorization", {
  enumerable: true,
  get: function () { return auth_js.refreshAuthorization; }
});
Object.defineProperty(exports, "registerClient", {
  enumerable: true,
  get: function () { return auth_js.registerClient; }
});
Object.defineProperty(exports, "selectResourceURL", {
  enumerable: true,
  get: function () { return auth_js.selectResourceURL; }
});
Object.defineProperty(exports, "startAuthorization", {
  enumerable: true,
  get: function () { return auth_js.startAuthorization; }
});
exports.InMemoryOAuthStorage = InMemoryOAuthStorage;
exports.InternalMastraMCPClient = InternalMastraMCPClient;
exports.MCPClient = MCPClient;
exports.MCPOAuthClientProvider = MCPOAuthClientProvider;
exports.MCPServer = MCPServer;
exports.createIntrospectionValidator = createIntrospectionValidator;
exports.createOAuthMiddleware = createOAuthMiddleware;
exports.createSimpleTokenProvider = createSimpleTokenProvider;
exports.createStaticTokenValidator = createStaticTokenValidator;
exports.extractBearerToken = extractBearerToken;
exports.generateProtectedResourceMetadata = generateProtectedResourceMetadata;
exports.generateWWWAuthenticateHeader = generateWWWAuthenticateHeader;
