// @mastra/evals/dist/chunk-OEOE7ZHN.js
import { RequestContext } from '@mastra/core/request-context';

// src/scorers/utils.ts
function getTextContentFromMastraDBMessage(message) {
  if (typeof message.content.content === "string" && message.content.content !== "") {
    return message.content.content;
  }
  if (message.content.parts && Array.isArray(message.content.parts)) {
    const textParts = message.content.parts.filter((p) => p.type === "text");
    return textParts.length > 0 ? textParts[textParts.length - 1]?.text || "" : "";
  }
  return "";
}
var roundToTwoDecimals = (num) => {
  return Math.round((num + Number.EPSILON) * 100) / 100;
};
function isCloserTo(value, target1, target2) {
  return Math.abs(value - target1) < Math.abs(value - target2);
}
var createTestRun = (input, output, additionalContext, requestContext) => {
  return {
    input: [{ role: "user", content: input }],
    output: { role: "assistant", text: output },
    additionalContext: additionalContext ?? {},
    requestContext: requestContext ?? {}
  };
};
var getUserMessageFromRunInput = (input) => {
  const message = input?.inputMessages.find(({ role }) => role === "user");
  return message ? getTextContentFromMastraDBMessage(message) : void 0;
};
var getSystemMessagesFromRunInput = (input) => {
  const systemMessages = [];
  if (input?.systemMessages) {
    systemMessages.push(
      ...input.systemMessages.map((msg) => {
        if (typeof msg.content === "string") {
          return msg.content;
        } else if (Array.isArray(msg.content)) {
          return msg.content.filter((part) => part.type === "text").map((part) => part.text || "").join(" ");
        }
        return "";
      }).filter((content) => content)
    );
  }
  if (input?.taggedSystemMessages) {
    Object.values(input.taggedSystemMessages).forEach((messages) => {
      messages.forEach((msg) => {
        if (typeof msg.content === "string") {
          systemMessages.push(msg.content);
        }
      });
    });
  }
  return systemMessages;
};
var getCombinedSystemPrompt = (input) => {
  const systemMessages = getSystemMessagesFromRunInput(input);
  return systemMessages.join("\n\n");
};
var getAssistantMessageFromRunOutput = (output) => {
  const message = output?.find(({ role }) => role === "assistant");
  return message ? getTextContentFromMastraDBMessage(message) : void 0;
};
var getReasoningFromRunOutput = (output) => {
  if (!output) return void 0;
  const message = output.find(({ role }) => role === "assistant");
  if (!message) return void 0;
  if (message.content.reasoning) {
    return message.content.reasoning;
  }
  const reasoningParts = message.content.parts?.filter((p) => p.type === "reasoning");
  if (reasoningParts && reasoningParts.length > 0) {
    const reasoningTexts = reasoningParts.map((p) => {
      if (p.details && Array.isArray(p.details)) {
        return p.details.filter((d) => d.type === "text").map((d) => d.text).join("");
      }
      return p.reasoning || "";
    }).filter(Boolean);
    return reasoningTexts.length > 0 ? reasoningTexts.join("\n") : void 0;
  }
  return void 0;
};
var createToolInvocation = ({
  toolCallId,
  toolName,
  args,
  result,
  state = "result"
}) => {
  return {
    toolCallId,
    toolName,
    args,
    result,
    state
  };
};
function createTestMessage({
  content,
  role,
  id = "test-message",
  toolInvocations = []
}) {
  return {
    id,
    role,
    content: {
      format: 2,
      parts: [{ type: "text", text: content }],
      content,
      ...toolInvocations.length > 0 && {
        toolInvocations: toolInvocations.map((ti) => ({
          toolCallId: ti.toolCallId,
          toolName: ti.toolName,
          args: ti.args,
          result: ti.result,
          state: ti.state
        }))
      }
    },
    createdAt: /* @__PURE__ */ new Date()
  };
}
var createAgentTestRun = ({
  inputMessages = [],
  output,
  rememberedMessages = [],
  systemMessages = [],
  taggedSystemMessages = {},
  requestContext = new RequestContext(),
  runId = crypto.randomUUID()
}) => {
  return {
    input: {
      inputMessages,
      rememberedMessages,
      systemMessages,
      taggedSystemMessages
    },
    output,
    requestContext,
    runId
  };
};
function extractToolCalls(output) {
  const toolCalls = [];
  const toolCallInfos = [];
  for (let messageIndex = 0; messageIndex < output.length; messageIndex++) {
    const message = output[messageIndex];
    if (message?.content?.toolInvocations) {
      for (let invocationIndex = 0; invocationIndex < message.content.toolInvocations.length; invocationIndex++) {
        const invocation = message.content.toolInvocations[invocationIndex];
        if (invocation && invocation.toolName && (invocation.state === "result" || invocation.state === "call")) {
          toolCalls.push(invocation.toolName);
          toolCallInfos.push({
            toolName: invocation.toolName,
            toolCallId: invocation.toolCallId || `${messageIndex}-${invocationIndex}`,
            messageIndex,
            invocationIndex
          });
        }
      }
    }
  }
  return { tools: toolCalls, toolCallInfos };
}
var extractInputMessages = (runInput) => {
  return runInput?.inputMessages?.map((msg) => getTextContentFromMastraDBMessage(msg)) || [];
};
var extractAgentResponseMessages = (runOutput) => {
  return runOutput.filter((msg) => msg.role === "assistant").map((msg) => getTextContentFromMastraDBMessage(msg));
};
function extractToolResults(output) {
  const results = [];
  for (const message of output) {
    const toolInvocations = message?.content?.toolInvocations;
    if (!toolInvocations) continue;
    for (const invocation of toolInvocations) {
      if (invocation.state === "result" && invocation.result !== void 0) {
        results.push({
          toolName: invocation.toolName,
          toolCallId: invocation.toolCallId || "",
          args: invocation.args || {},
          result: invocation.result
        });
      }
    }
  }
  return results;
}

export { createAgentTestRun, createTestMessage, createTestRun, createToolInvocation, extractAgentResponseMessages, extractInputMessages, extractToolCalls, extractToolResults, getAssistantMessageFromRunOutput, getCombinedSystemPrompt, getReasoningFromRunOutput, getSystemMessagesFromRunInput, getTextContentFromMastraDBMessage, getUserMessageFromRunInput, isCloserTo, roundToTwoDecimals };
